// @midscene/ios/dist/es/cli.mjs
import { createReportCliCommands, getMidsceneLocationSchema, z } from "@midscene/core";
import { reportCLIError, runToolsCLI } from "@midscene/shared/cli";
import { getDebug } from "@midscene/shared/logger";
import { BaseMidsceneTools } from "@midscene/shared/mcp/base-tools";
import { Agent } from "@midscene/core/agent";
import { MIDSCENE_IOS_DEVICE_CLASS_OVERRIDE } from "@midscene/shared/env";
import { mergeAndNormalizeAppNameMapping, normalizeForComparison } from "@midscene/shared/utils";
import node_assert from "node:assert";
import { defineAction, defineActionClearInput, defineActionCursorMove, defineActionDoubleClick, defineActionDragAndDrop, defineActionKeyboardPress, defineActionLongPress, defineActionPinch, defineActionScroll, defineActionSwipe, defineActionTap, normalizeMobileSwipeParam, normalizePinchParam } from "@midscene/core/device";
import { sleep } from "@midscene/core/utils";
import { DEFAULT_WDA_PORT } from "@midscene/shared/constants";
import { createImgBase64ByFormat } from "@midscene/shared/img";
import { WDAManager, WebDriverClient } from "@midscene/webdriver";
const defaultAppNameMapping = {
    微信: 'com.tencent.xin',
    企业微信: 'com.tencent.ww',
    微信读书: 'com.tencent.weread',
    微信听书: 'com.tencent.wehear',
    QQ: 'com.tencent.mqq',
    QQ音乐: 'com.tencent.QQMusic',
    QQ阅读: 'com.tencent.qqreaderiphone',
    QQ邮箱: 'com.tencent.qqmail',
    QQ浏览器: 'com.tencent.mttlite',
    TIM: 'com.tencent.tim',
    微视: 'com.tencent.microvision',
    腾讯新闻: 'com.tencent.info',
    腾讯视频: 'com.tencent.live4iphone',
    腾讯动漫: 'com.tencent.ied.app.comic',
    腾讯微云: 'com.tencent.weiyun',
    腾讯体育: 'com.tencent.sportskbs',
    腾讯文档: 'com.tencent.txdocs',
    腾讯翻译君: 'com.tencent.qqtranslator',
    腾讯课堂: 'com.tencent.edu',
    腾讯地图: 'com.tencent.sosomap',
    小鹅拼拼: 'com.tencent.dwdcoco',
    全民k歌: 'com.tencent.QQKSong',
    支付宝: 'com.alipay.iphoneclient',
    钉钉: 'com.laiwang.DingTalk',
    闲鱼: 'com.taobao.fleamarket',
    淘宝: 'com.taobao.taobao4iphone',
    斗鱼: 'tv.douyu.live',
    天猫: 'com.taobao.tmall',
    口碑: 'com.taobao.kbmeishi',
    饿了么: 'me.ele.ios.eleme',
    高德地图: 'com.autonavi.amap',
    UC浏览器: 'com.ucweb.iphone.lowversion',
    一淘: 'com.taobao.etaocoupon',
    飞猪: 'com.taobao.travel',
    虾米音乐: 'com.xiami.spark',
    淘票票: 'com.taobao.movie.MoviePhoneClient',
    优酷: 'com.youku.YouKu',
    菜鸟裹裹: 'com.cainiao.cnwireless',
    土豆视频: 'com.tudou.tudouiphone',
    抖音: 'com.ss.iphone.ugc.Aweme',
    抖音极速版: 'com.ss.iphone.ugc.aweme.lite',
    抖音火山版: 'com.ss.iphone.ugc.Live',
    懂车帝: 'com.ss.ios.auto',
    Tiktok: 'com.zhiliaoapp.musically',
    飞书: 'com.bytedance.ee.lark',
    今日头条: 'com.ss.iphone.article.News',
    西瓜视频: 'com.ss.iphone.article.Video',
    皮皮虾: 'com.bd.iphone.super',
    美团: 'com.meituan.imeituan',
    美团外卖: 'com.meituan.itakeaway',
    大众点评: 'com.dianping.dpscope',
    美团优选: 'com.meituan.iyouxuan',
    美团优选团长: 'com.meituan.igrocery.gh',
    美团骑手: 'com.meituan.banma.homebrew',
    美团开店宝: 'com.meituan.imerchantbiz',
    美团拍店: 'com.meituan.pai',
    美团众包: 'com.meituan.banma.crowdsource',
    美团买菜: 'com.baobaoaichi.imaicai',
    京东: 'com.360buy.jdmobile',
    京东读书: 'com.jd.reader',
    网易新闻: 'com.netease.news',
    网易云音乐: 'com.netease.cloudmusic',
    网易邮箱大师: 'com.netease.macmail',
    网易严选: 'com.netease.yanxuan',
    网易公开课: 'com.netease.videoHD',
    网易有道词典: 'youdaoPro',
    有道云笔记: 'com.youdao.note.YoudaoNoteMac',
    百度: 'com.baidu.BaiduMobile',
    百度网盘: 'com.baidu.netdisk',
    百度贴吧: 'com.baidu.tieba',
    百度地图: 'com.baidu.map',
    百度阅读: 'com.baidu.yuedu',
    百度翻译: 'com.baidu.translate',
    百度文库: 'com.baidu.Wenku',
    百度视频: 'com.baidu.videoiphone',
    百度输入法: 'com.baidu.inputMethod',
    快手: 'com.jiangjia.gif',
    快手极速版: 'com.kuaishou.nebula',
    哔哩哔哩: 'tv.danmaku.bilianime',
    芒果TV: 'com.hunantv.imgotv',
    苏宁易购: 'SuningEMall',
    微博: 'com.sina.weibo',
    微博极速版: 'com.sina.weibolite',
    微博国际: 'com.weibo.international',
    墨客: 'com.moke.moke.iphone',
    豆瓣: 'com.douban.frodo',
    知乎: 'com.zhihu.ios',
    小红书: 'com.xingin.discover',
    喜马拉雅: 'com.gemd.iting',
    得到: 'com.luojilab.LuoJiFM-IOS',
    得物: 'com.siwuai.duapp',
    起点读书: 'm.qidian.QDReaderAppStore',
    番茄小说: 'com.dragon.read',
    书旗小说: 'com.shuqicenter.reader',
    拼多多: 'com.xunmeng.pinduoduo',
    多点: 'com.dmall.dmall',
    便利蜂: 'com.bianlifeng.customer.ios',
    亿通行: 'com.ruubypay.yitongxing',
    云闪付: 'com.unionpay.chsp',
    大都会Metro: 'com.DDH.SHSubway',
    爱奇艺视频: 'com.qiyi.iphone',
    搜狐视频: 'com.sohu.iPhoneVideo',
    搜狐新闻: 'com.sohu.newspaper',
    搜狗浏览器: 'com.sogou.SogouExplorerMobile',
    虎牙: 'com.yy.kiwi',
    比心: 'com.yitan.bixin',
    转转: 'com.wuba.zhuanzhuan',
    YY: 'yyvoice',
    绿洲: 'com.sina.oasis',
    陌陌: 'com.wemomo.momoappdemo1',
    什么值得买: 'com.smzdm.client.ios',
    美团秀秀: 'com.meitu.mtxx',
    唯品会: 'com.vipshop.iphone',
    唱吧: 'com.changba.ktv',
    酷狗音乐: 'com.kugou.kugou1002',
    CSDN: 'net.csdn.CsdnPlus',
    多抓鱼: 'com.duozhuyu.dejavu',
    自如: 'com.ziroom.ZiroomProject',
    携程: 'ctrip.com',
    去哪儿旅行: 'com.qunar.iphoneclient8',
    Xmind: 'net.xmind.brownieapp',
    印象笔记: 'com.yinxiang.iPhone',
    欧陆词典: 'eusoft.eudic.pro',
    115: 'com.115.personal',
    名片全能王: 'com.intsig.camcard.lite',
    中国银行: 'com.boc.BOCMBCI',
    '58同城': 'com.taofang.iphone',
    'Google Chrome': 'com.google.chrome.ios',
    Gmail: 'com.google.Gmail',
    Facebook: 'com.facebook.Facebook',
    Firefox: 'org.mozilla.ios.Firefox',
    Messenger: 'com.facebook.Messenger',
    Instagram: 'com.burbn.instagram',
    Starbucks: 'com.starbucks.mystarbucks',
    'Luckin Coffee': 'com.bjlc.luckycoffee',
    Line: 'jp.naver.line',
    Linkedin: 'com.linkedin.LinkedIn',
    Dcard: 'com.dcard.app.Dcard',
    Youtube: 'com.google.ios.youtube',
    Spotify: 'com.spotify.client',
    Netflix: 'com.netflix.Netflix',
    Twitter: 'com.atebits.Tweetie2',
    WhatsApp: 'net.whatsapp.WhatsApp',
    Safari: 'com.apple.mobilesafari',
    'App Store': 'com.apple.AppStore',
    设置: 'com.apple.Preferences',
    相机: 'com.apple.camera',
    照片: 'com.apple.mobileslideshow',
    时钟: 'com.apple.mobiletimer',
    闹钟: 'com.apple.mobiletimer',
    备忘录: 'com.apple.mobilenotes',
    提醒事项: 'com.apple.reminders',
    快捷指令: 'com.apple.shortcuts',
    天气: 'com.apple.weather',
    日历: 'com.apple.mobilecal',
    地图: 'com.apple.Maps',
    电话: 'com.apple.mobilephone',
    通讯录: 'com.apple.MobileAddressBook',
    信息: 'com.apple.MobileSMS',
    FaceTime: 'com.apple.facetime',
    计算器: 'com.apple.calculator',
    家庭: 'com.apple.Home',
    健康: 'com.apple.Health',
    钱包: 'com.apple.Passbook',
    股市: 'com.apple.stocks',
    图书: 'com.apple.iBooks',
    新闻: 'com.apple.news',
    视频: 'com.apple.tv',
    文件: 'com.apple.DocumentsApp',
    邮件: 'com.apple.mobilemail',
    查找: 'com.apple.findmy',
    翻译: 'com.apple.Translate',
    音乐: 'com.apple.Music',
    播客: 'com.apple.podcasts',
    库乐队: 'com.apple.mobilegarageband',
    语音备忘录: 'com.apple.VoiceMemos',
    iMovie: 'com.apple.iMovie',
    Watch: 'com.apple.Bridge',
    'Apple Store': 'com.apple.store.Jolly',
    TestFlight: 'com.apple.TestFlight',
    Keynote: 'com.apple.Keynote',
    'Keynote 讲演': 'com.apple.Keynote'
};
const debugIOS = getDebug('webdriver:ios');
const WDA_MJPEG_SCREENSHOT_QUALITY = 50;
const WDA_MJPEG_FRAMERATE = 30;
const WDA_MJPEG_SCALING_FACTOR = 50;
class IOSWebDriverClient extends WebDriverClient {
    async launchApp(bundleId) {
        this.ensureSession();
        try {
            await this.makeRequest('POST', `/session/${this.sessionId}/wda/apps/launch`, {
                bundleId
            });
            debugIOS(`Launched app: ${bundleId}`);
        } catch (error) {
            debugIOS(`Failed to launch app ${bundleId}: ${error}`);
            throw error;
        }
    }
    async activateApp(bundleId) {
        this.ensureSession();
        await this.makeRequest('POST', `/session/${this.sessionId}/wda/apps/activate`, {
            bundleId
        });
    }
    async terminateApp(bundleId) {
        this.ensureSession();
        await this.makeRequest('POST', `/session/${this.sessionId}/wda/apps/terminate`, {
            bundleId
        });
    }
    async openUrl(url) {
        this.ensureSession();
        try {
            await this.makeRequest('POST', `/session/${this.sessionId}/url`, {
                url
            });
        } catch (error) {
            debugIOS(`Direct URL opening failed, trying Safari fallback: ${error}`);
            await this.launchApp('com.apple.mobilesafari');
            await new Promise((resolve)=>setTimeout(resolve, 2000));
            await this.makeRequest('POST', `/session/${this.sessionId}/url`, {
                url
            });
        }
    }
    async pressHomeButton() {
        this.ensureSession();
        try {
            await this.makeRequest('POST', `/session/${this.sessionId}/wda/pressButton`, {
                name: 'home'
            });
            debugIOS('Home button pressed using hardware key');
        } catch (error) {
            debugIOS(`Failed to press home button: ${error}`);
            throw new Error(`Failed to press home button: ${error}`);
        }
    }
    async appSwitcher() {
        this.ensureSession();
        try {
            const windowSize = await this.getWindowSize();
            debugIOS('Triggering app switcher with slow swipe up gesture');
            const centerX = windowSize.width / 2;
            const startY = windowSize.height - 5;
            const endY = 0.5 * windowSize.height;
            await this.swipe(centerX, startY, centerX, endY, 1500);
            await new Promise((resolve)=>setTimeout(resolve, 800));
        } catch (error) {
            debugIOS(`App switcher failed: ${error}`);
            throw new Error(`Failed to trigger app switcher: ${error}`);
        }
    }
    async pressKey(key) {
        this.ensureSession();
        debugIOS(`Attempting to press key: ${key}`);
        if ('Enter' === key || 'Return' === key || 'return' === key) {
            debugIOS('Handling Enter/Return key for iOS');
            await this.makeRequest('POST', `/session/${this.sessionId}/wda/keys`, {
                value: [
                    '\n'
                ]
            });
            debugIOS('Sent newline character for Enter key');
            await new Promise((resolve)=>setTimeout(resolve, 100));
            return;
        }
        if ('Backspace' === key || 'Delete' === key) try {
            await this.makeRequest('POST', `/session/${this.sessionId}/wda/keys`, {
                value: [
                    '\b'
                ]
            });
            debugIOS('Sent backspace character');
            return;
        } catch (error) {
            debugIOS(`Backspace failed: ${error}`);
        }
        if ('Space' === key) try {
            await this.makeRequest('POST', `/session/${this.sessionId}/wda/keys`, {
                value: [
                    ' '
                ]
            });
            debugIOS('Sent space character');
            return;
        } catch (error) {
            debugIOS(`Space key failed: ${error}`);
        }
        const normalizedKey = this.normalizeKeyName(key);
        const iosKeyMap = {
            Tab: '\t',
            ArrowUp: '\uE013',
            ArrowDown: '\uE015',
            ArrowLeft: '\uE012',
            ArrowRight: '\uE014',
            Home: '\uE011',
            End: '\uE010'
        };
        if (iosKeyMap[normalizedKey]) try {
            await this.makeRequest('POST', `/session/${this.sessionId}/wda/keys`, {
                value: [
                    iosKeyMap[normalizedKey]
                ]
            });
            debugIOS(`Sent WebDriver key code for: ${key}`);
            return;
        } catch (error) {
            debugIOS(`WebDriver key failed for "${key}": ${error}`);
        }
        if (1 === key.length) try {
            await this.makeRequest('POST', `/session/${this.sessionId}/wda/keys`, {
                value: [
                    key
                ]
            });
            debugIOS(`Sent single character: "${key}"`);
            return;
        } catch (error) {
            debugIOS(`Failed to send character "${key}": ${error}`);
        }
        debugIOS(`Warning: Key "${key}" is not supported on iOS platform`);
        throw new Error(`Key "${key}" is not supported on iOS platform`);
    }
    async getActiveElement() {
        this.ensureSession();
        debugIOS('Getting active element');
        try {
            const response = await this.makeRequest('GET', `/session/${this.sessionId}/element/active`);
            const elementId = response.value?.ELEMENT || response.value?.['element-6066-11e4-a52e-4f735466cecf'] || response.ELEMENT || response['element-6066-11e4-a52e-4f735466cecf'];
            if (elementId) {
                debugIOS(`Got active element ID: ${elementId}`);
                return elementId;
            }
            debugIOS('No active element found');
            return null;
        } catch (error) {
            debugIOS(`Failed to get active element: ${error}`);
            return null;
        }
    }
    async clearElement(elementId) {
        this.ensureSession();
        debugIOS(`Clearing element: ${elementId}`);
        try {
            await this.makeRequest('POST', `/session/${this.sessionId}/element/${elementId}/clear`);
            debugIOS('Element cleared successfully');
        } catch (error) {
            debugIOS(`Failed to clear element: ${error}`);
            throw new Error(`Failed to clear element: ${error}`);
        }
    }
    async clearActiveElement() {
        try {
            const elementId = await this.getActiveElement();
            if (!elementId) {
                debugIOS('No active element to clear');
                return false;
            }
            await this.clearElement(elementId);
            return true;
        } catch (error) {
            debugIOS(`Failed to clear active element: ${error}`);
            return false;
        }
    }
    normalizeKeyName(key) {
        return key.charAt(0).toUpperCase() + key.slice(1).toLowerCase();
    }
    async dismissKeyboard(keyNames) {
        this.ensureSession();
        try {
            await this.makeRequest('POST', `/session/${this.sessionId}/wda/keyboard/dismiss`, {
                keyNames: keyNames || [
                    'done'
                ]
            });
            debugIOS('Dismissed keyboard using WDA API');
            return true;
        } catch (error) {
            debugIOS(`Failed to dismiss keyboard: ${error}`);
            return false;
        }
    }
    async typeText(text) {
        this.ensureSession();
        try {
            const cleanText = text.trim();
            await this.makeRequest('POST', `/session/${this.sessionId}/wda/keys`, {
                value: cleanText.split('')
            });
            debugIOS(`Typed text: "${text}"`);
        } catch (error) {
            debugIOS(`Failed to type text "${text}": ${error}`);
            throw new Error(`Failed to type text: ${error}`);
        }
    }
    async tap(x, y) {
        this.ensureSession();
        try {
            await this.makeRequest('POST', `/session/${this.sessionId}/wda/tap`, {
                x,
                y
            });
            debugIOS(`Tapped at coordinates (${x}, ${y})`);
        } catch (error) {
            debugIOS(`New tap endpoint failed, trying legacy endpoint: ${error}`);
            try {
                await this.makeRequest('POST', `/session/${this.sessionId}/wda/tap/0`, {
                    x,
                    y
                });
                debugIOS(`Tapped at coordinates (${x}, ${y}) using legacy endpoint`);
            } catch (fallbackError) {
                debugIOS(`Failed to tap at (${x}, ${y}): ${fallbackError}`);
                throw new Error(`Failed to tap at coordinates: ${fallbackError}`);
            }
        }
    }
    async swipe(fromX, fromY, toX, toY, duration = 500) {
        this.ensureSession();
        const actions = {
            actions: [
                {
                    type: 'pointer',
                    id: 'finger1',
                    parameters: {
                        pointerType: 'touch'
                    },
                    actions: [
                        {
                            type: 'pointerMove',
                            duration: 0,
                            x: fromX,
                            y: fromY
                        },
                        {
                            type: 'pointerDown',
                            button: 0
                        },
                        {
                            type: 'pause',
                            duration: 100
                        },
                        {
                            type: 'pointerMove',
                            duration,
                            x: toX,
                            y: toY
                        },
                        {
                            type: 'pointerUp',
                            button: 0
                        }
                    ]
                }
            ]
        };
        await this.makeRequest('POST', `/session/${this.sessionId}/actions`, actions);
        debugIOS(`Swiped using W3C Actions from (${fromX}, ${fromY}) to (${toX}, ${toY}) in ${duration}ms`);
    }
    async pinch(centerX, centerY, startDistance, endDistance, duration = 500) {
        this.ensureSession();
        const halfStart = startDistance / 2;
        const halfEnd = endDistance / 2;
        const actions = {
            actions: [
                {
                    type: 'pointer',
                    id: 'finger1',
                    parameters: {
                        pointerType: 'touch'
                    },
                    actions: [
                        {
                            type: 'pointerMove',
                            duration: 0,
                            x: centerX,
                            y: Math.round(centerY - halfStart)
                        },
                        {
                            type: 'pointerDown',
                            button: 0
                        },
                        {
                            type: 'pause',
                            duration: 100
                        },
                        {
                            type: 'pointerMove',
                            duration,
                            x: centerX,
                            y: Math.round(centerY - halfEnd)
                        },
                        {
                            type: 'pointerUp',
                            button: 0
                        }
                    ]
                },
                {
                    type: 'pointer',
                    id: 'finger2',
                    parameters: {
                        pointerType: 'touch'
                    },
                    actions: [
                        {
                            type: 'pointerMove',
                            duration: 0,
                            x: centerX,
                            y: Math.round(centerY + halfStart)
                        },
                        {
                            type: 'pointerDown',
                            button: 0
                        },
                        {
                            type: 'pause',
                            duration: 100
                        },
                        {
                            type: 'pointerMove',
                            duration,
                            x: centerX,
                            y: Math.round(centerY + halfEnd)
                        },
                        {
                            type: 'pointerUp',
                            button: 0
                        }
                    ]
                }
            ]
        };
        await this.makeRequest('POST', `/session/${this.sessionId}/actions`, actions);
        debugIOS(`Pinched at (${centerX}, ${centerY}) from distance ${startDistance} to ${endDistance} in ${duration}ms`);
    }
    async longPress(x, y, duration = 1000) {
        this.ensureSession();
        await this.makeRequest('POST', `/session/${this.sessionId}/wda/touchAndHold`, {
            x,
            y,
            duration: duration / 1000
        });
        debugIOS(`Long pressed at coordinates (${x}, ${y}) for ${duration}ms`);
    }
    async doubleTap(x, y) {
        this.ensureSession();
        await this.makeRequest('POST', `/session/${this.sessionId}/wda/doubleTap`, {
            x,
            y
        });
        debugIOS(`Double tapped at coordinates (${x}, ${y})`);
    }
    async tripleTap(x, y) {
        this.ensureSession();
        await this.makeRequest('POST', `/session/${this.sessionId}/wda/tapWithNumberOfTaps`, {
            x,
            y,
            numberOfTaps: 3,
            numberOfTouches: 1
        });
        debugIOS(`Triple tapped at coordinates (${x}, ${y})`);
    }
    async getScreenScale() {
        this.ensureSession();
        try {
            const screenResponse = await this.makeRequest('GET', `/session/${this.sessionId}/wda/screen`);
            if (screenResponse?.value?.scale) {
                debugIOS(`Got screen scale from WDA screen endpoint: ${screenResponse.value.scale}`);
                return screenResponse.value.scale;
            }
        } catch (error) {
            debugIOS(`Failed to get screen scale from /wda/screen: ${error}`);
        }
        try {
            debugIOS('Calculating screen scale from screenshot and window size');
            const [screenshotBase64, windowSize] = await Promise.all([
                this.takeScreenshot(),
                this.getWindowSize()
            ]);
            const { imageInfoOfBase64 } = await import("@midscene/shared/img");
            const { width: screenshotWidth, height: screenshotHeight } = await imageInfoOfBase64(screenshotBase64);
            const scale = Math.max(screenshotWidth, screenshotHeight) / Math.max(windowSize.width, windowSize.height);
            const roundedScale = Math.round(scale);
            debugIOS(`Calculated screen scale: ${roundedScale} (screenshot: ${screenshotWidth}x${screenshotHeight}, window: ${windowSize.width}x${windowSize.height})`);
            return roundedScale;
        } catch (error) {
            debugIOS(`Failed to calculate screen scale: ${error}`);
        }
        debugIOS('No screen scale found');
        return null;
    }
    async createSession(capabilities) {
        const defaultCapabilities = {
            platformName: 'iOS',
            automationName: 'XCUITest',
            shouldUseSingletonTestManager: false,
            shouldUseTestManagerForVisibilityDetection: false,
            ...capabilities
        };
        const session = await super.createSession(defaultCapabilities);
        await this.setupIOSSession();
        return session;
    }
    async setupIOSSession() {
        if (!this.sessionId) return;
        try {
            await this.makeRequest('POST', `/session/${this.sessionId}/appium/settings`, {
                snapshotMaxDepth: 50,
                elementResponseAttributes: 'type,label,name,value,rect,enabled,visible',
                mjpegServerScreenshotQuality: WDA_MJPEG_SCREENSHOT_QUALITY,
                mjpegServerFramerate: WDA_MJPEG_FRAMERATE,
                mjpegScalingFactor: WDA_MJPEG_SCALING_FACTOR
            });
            debugIOS('iOS session configuration applied (including MJPEG settings)');
        } catch (error) {
            debugIOS(`Failed to apply iOS session configuration: ${error}`);
        }
    }
    async executeRequest(method, endpoint, data) {
        return this.makeRequest(method, this.buildSessionEndpoint(endpoint), data);
    }
}
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
const debugDevice = getDebug('ios:device');
const iosInputParamSchema = z.object({
    value: z.string().describe('The text to input. Provide the final content for replace/append modes, or an empty string when using clear mode to remove existing text.'),
    autoDismissKeyboard: z.boolean().optional().describe('Whether to dismiss the keyboard after input. Defaults to true if not specified. Set to false to keep the keyboard visible after input.'),
    mode: z.preprocess((val)=>'append' === val ? 'typeOnly' : val, z["enum"]([
        'replace',
        'clear',
        'typeOnly'
    ]).default('replace').optional().describe('Input mode: "replace" (default) - clear the field and input the value; "typeOnly" - type the value directly without clearing the field first; "clear" - clear the field without inputting new text.')),
    locate: getMidsceneLocationSchema().describe('The input field to be filled').optional()
});
const WDA_HTTP_METHODS = [
    'GET',
    'POST',
    'DELETE',
    'PUT'
];
const DEFAULT_WDA_MJPEG_PORT = 9100;
class IOSDevice {
    actionSpace() {
        const defaultActions = [
            defineActionTap(async (param)=>{
                const element = param.locate;
                node_assert(element, 'Element not found, cannot tap');
                await this.mouseClick(element.center[0], element.center[1]);
            }),
            defineActionDoubleClick(async (param)=>{
                const element = param.locate;
                node_assert(element, 'Element not found, cannot double click');
                await this.doubleTap(element.center[0], element.center[1]);
            }),
            defineAction({
                name: 'Input',
                description: 'Input text into the input field',
                interfaceAlias: 'aiInput',
                paramSchema: iosInputParamSchema,
                sample: {
                    value: 'test@example.com',
                    locate: {
                        prompt: 'the email input field'
                    }
                },
                call: async (param)=>{
                    const element = param.locate;
                    if ('typeOnly' !== param.mode) await this.clearInput(element);
                    if ('clear' === param.mode) return;
                    if (!param || !param.value) return;
                    const autoDismissKeyboard = param.autoDismissKeyboard ?? this.options?.autoDismissKeyboard;
                    await this.typeText(param.value, {
                        autoDismissKeyboard
                    });
                }
            }),
            defineActionScroll(async (param)=>{
                const element = param.locate;
                const startingPoint = element ? {
                    left: element.center[0],
                    top: element.center[1]
                } : void 0;
                const scrollToEventName = param?.scrollType;
                if ('scrollToTop' === scrollToEventName) await this.scrollUntilTop(startingPoint);
                else if ('scrollToBottom' === scrollToEventName) await this.scrollUntilBottom(startingPoint);
                else if ('scrollToRight' === scrollToEventName) await this.scrollUntilRight(startingPoint);
                else if ('scrollToLeft' === scrollToEventName) await this.scrollUntilLeft(startingPoint);
                else if ('singleAction' !== scrollToEventName && scrollToEventName) throw new Error(`Unknown scroll event type: ${scrollToEventName}, param: ${JSON.stringify(param)}`);
                else {
                    if (param?.direction !== 'down' && param && param.direction) if ('up' === param.direction) await this.scrollUp(param.distance || void 0, startingPoint);
                    else if ('left' === param.direction) await this.scrollLeft(param.distance || void 0, startingPoint);
                    else if ('right' === param.direction) await this.scrollRight(param.distance || void 0, startingPoint);
                    else throw new Error(`Unknown scroll direction: ${param.direction}`);
                    else await this.scrollDown(param?.distance || void 0, startingPoint);
                    await sleep(500);
                }
            }),
            defineActionDragAndDrop(async (param)=>{
                const from = param.from;
                const to = param.to;
                node_assert(from, 'missing "from" param for drag and drop');
                node_assert(to, 'missing "to" param for drag and drop');
                await this.swipe(from.center[0], from.center[1], to.center[0], to.center[1], 1000);
            }),
            defineActionSwipe(async (param)=>{
                const { startPoint, endPoint, duration, repeatCount } = normalizeMobileSwipeParam(param, await this.size());
                for(let i = 0; i < repeatCount; i++)await this.swipe(startPoint.x, startPoint.y, endPoint.x, endPoint.y, duration);
            }),
            defineActionKeyboardPress(async (param)=>{
                await this.pressKey(param.keyName);
            }),
            defineActionCursorMove(async (param)=>{
                const arrowKey = 'left' === param.direction ? 'ArrowLeft' : 'ArrowRight';
                const times = param.times ?? 1;
                for(let i = 0; i < times; i++){
                    await this.pressKey(arrowKey);
                    await sleep(100);
                }
            }),
            defineActionLongPress(async (param)=>{
                const element = param.locate;
                node_assert(element, 'LongPress requires an element to be located');
                const [x, y] = element.center;
                await this.longPress(x, y, param?.duration);
            }),
            defineActionPinch(async (param)=>{
                const { centerX, centerY, startDistance, endDistance, duration } = normalizePinchParam(param, await this.size());
                await this.wdaBackend.pinch(centerX, centerY, startDistance, endDistance, duration);
            }),
            defineActionClearInput(async (param)=>{
                await this.clearInput(param.locate);
            })
        ];
        const platformSpecificActions = Object.values(createPlatformActions(this));
        const customActions = this.customActions || [];
        return [
            ...defaultActions,
            ...platformSpecificActions,
            ...customActions
        ];
    }
    describe() {
        return this.description || `Device ID: ${this.deviceId}`;
    }
    async getConnectedDeviceInfo() {
        return await this.wdaBackend.getDeviceInfo();
    }
    async connect() {
        node_assert(!this.destroyed, `IOSDevice ${this.deviceId} has been destroyed and cannot execute commands`);
        debugDevice(`Connecting to iOS device: ${this.deviceId}`);
        try {
            await this.wdaManager.start();
            await this.wdaBackend.createSession();
            const deviceInfo = await this.wdaBackend.getDeviceInfo();
            if (deviceInfo?.udid) {
                this.deviceId = deviceInfo.udid;
                debugDevice(`Updated device ID to real UDID: ${this.deviceId}`);
            }
            const size = await this.getScreenSize();
            this.description = `
UDID: ${this.deviceId}${deviceInfo ? `
Name: ${deviceInfo.name}
Model: ${deviceInfo.model}` : ''}
Type: WebDriverAgent
ScreenSize: ${size.width}x${size.height} (DPR: ${size.scale})
`;
            debugDevice('iOS device connected successfully', this.description);
        } catch (e) {
            debugDevice(`Failed to connect to iOS device: ${e}`);
            throw new Error(`Unable to connect to iOS device ${this.deviceId}: ${e}`);
        }
    }
    setAppNameMapping(mapping) {
        this.appNameMapping = mapping;
    }
    resolveBundleId(appName) {
        const normalizedAppName = normalizeForComparison(appName);
        return this.appNameMapping[normalizedAppName];
    }
    async launch(uri) {
        this.uri = uri;
        try {
            debugDevice(`Launching app: ${uri}`);
            if (uri.startsWith('http://') || uri.startsWith('https://') || uri.includes('://')) await this.openUrl(uri);
            else {
                const resolvedUri = this.resolveBundleId(uri) ?? uri;
                await this.wdaBackend.launchApp(resolvedUri);
            }
            debugDevice(`Successfully launched: ${uri}`);
        } catch (error) {
            debugDevice(`Error launching ${uri}: ${error}`);
            throw new Error(`Failed to launch ${uri}: ${error.message}`);
        }
        return this;
    }
    async terminate(bundleId) {
        const resolved = this.resolveBundleId(bundleId) ?? bundleId;
        try {
            debugDevice(`Terminating app: ${resolved}`);
            await this.wdaBackend.terminateApp(resolved);
            debugDevice(`Successfully terminated: ${resolved}`);
        } catch (error) {
            debugDevice(`Error terminating ${resolved}: ${error}`);
            throw new Error(`Failed to terminate ${resolved}: ${error.message}`);
        }
    }
    async getElementsInfo() {
        return [];
    }
    async getElementsNodeTree() {
        return {
            node: null,
            children: []
        };
    }
    async initializeDevicePixelRatio() {
        if (this.devicePixelRatioInitialized) return;
        const apiScale = await this.wdaBackend.getScreenScale();
        node_assert(apiScale && apiScale > 0, 'Failed to get device pixel ratio from WebDriverAgent API');
        debugDevice(`Got screen scale from WebDriverAgent API: ${apiScale}`);
        this.devicePixelRatio = apiScale;
        this.devicePixelRatioInitialized = true;
    }
    async getScreenSize() {
        await this.initializeDevicePixelRatio();
        const windowSize = await this.wdaBackend.getWindowSize();
        return {
            width: windowSize.width,
            height: windowSize.height,
            scale: this.devicePixelRatio
        };
    }
    async size() {
        const screenSize = await this.getScreenSize();
        return {
            width: screenSize.width,
            height: screenSize.height
        };
    }
    async screenshotBase64() {
        debugDevice('Taking screenshot via WDA');
        try {
            const base64Data = await this.wdaBackend.takeScreenshot();
            const result = createImgBase64ByFormat('png', base64Data);
            debugDevice('Screenshot taken successfully');
            return result;
        } catch (error) {
            debugDevice(`Screenshot failed: ${error}`);
            throw new Error(`Failed to take screenshot: ${error}`);
        }
    }
    async clearInput(element) {
        if (element) {
            await this.tap(element.center[0], element.center[1]);
            await sleep(100);
        }
        debugDevice('Attempting to clear input with WebDriver Clear API');
        const cleared = await this.wdaBackend.clearActiveElement();
        cleared ? debugDevice('Successfully cleared input with WebDriver Clear API') : debugDevice('WebDriver Clear API returned false (no active element or clear failed)');
    }
    async url() {
        return '';
    }
    async tap(x, y) {
        await this.wdaBackend.tap(Math.round(x), Math.round(y));
    }
    async mouseClick(x, y) {
        debugDevice(`mouseClick at coordinates (${x}, ${y})`);
        await this.tap(x, y);
    }
    async doubleTap(x, y) {
        await this.wdaBackend.doubleTap(Math.round(x), Math.round(y));
    }
    async tripleTap(x, y) {
        await this.wdaBackend.tripleTap(Math.round(x), Math.round(y));
    }
    async longPress(x, y, duration = 1000) {
        await this.wdaBackend.longPress(Math.round(x), Math.round(y), duration);
    }
    async swipe(fromX, fromY, toX, toY, duration = 500) {
        await this.wdaBackend.swipe(Math.round(fromX), Math.round(fromY), Math.round(toX), Math.round(toY), duration);
    }
    async typeText(text, options) {
        if (!text) return;
        const shouldAutoDismissKeyboard = options?.autoDismissKeyboard ?? this.options?.autoDismissKeyboard ?? true;
        debugDevice(`Typing text: "${text}"`);
        try {
            await sleep(200);
            await this.wdaBackend.typeText(text);
            await sleep(300);
        } catch (error) {
            debugDevice(`Failed to type text with WDA: ${error}`);
            throw error;
        }
        if (shouldAutoDismissKeyboard) await this.hideKeyboard();
    }
    async pressKey(key) {
        await this.wdaBackend.pressKey(key);
    }
    async scrollUp(distance, startPoint) {
        const { width, height } = await this.size();
        const start = startPoint ? {
            x: Math.round(startPoint.left),
            y: Math.round(startPoint.top)
        } : {
            x: Math.round(width / 2),
            y: Math.round(height / 2)
        };
        const scrollDistance = Math.round(distance || height / 3);
        await this.swipe(start.x, start.y, start.x, start.y + scrollDistance);
    }
    async scrollDown(distance, startPoint) {
        const { width, height } = await this.size();
        const start = startPoint ? {
            x: Math.round(startPoint.left),
            y: Math.round(startPoint.top)
        } : {
            x: Math.round(width / 2),
            y: Math.round(height / 2)
        };
        const scrollDistance = Math.round(distance || height / 3);
        await this.swipe(start.x, start.y, start.x, start.y - scrollDistance);
    }
    async scrollLeft(distance, startPoint) {
        const { width, height } = await this.size();
        const start = startPoint ? {
            x: Math.round(startPoint.left),
            y: Math.round(startPoint.top)
        } : {
            x: Math.round(width / 2),
            y: Math.round(height / 2)
        };
        const scrollDistance = Math.round(distance || 0.7 * width);
        await this.swipe(start.x, start.y, start.x + scrollDistance, start.y);
    }
    async scrollRight(distance, startPoint) {
        const { width, height } = await this.size();
        const start = startPoint ? {
            x: Math.round(startPoint.left),
            y: Math.round(startPoint.top)
        } : {
            x: Math.round(width / 2),
            y: Math.round(height / 2)
        };
        const scrollDistance = Math.round(distance || 0.7 * width);
        await this.swipe(start.x, start.y, start.x - scrollDistance, start.y);
    }
    async scrollUntilTop(startPoint) {
        debugDevice('Using screenshot-based scroll detection for better reliability');
        await this.scrollUntilBoundary('up', startPoint, 1);
    }
    async scrollUntilBottom(startPoint) {
        debugDevice('Using screenshot-based scroll detection for better reliability');
        await this.scrollUntilBoundary('down', startPoint, 1);
    }
    compareScreenshots(screenshot1, screenshot2, tolerancePercent = 2) {
        if (screenshot1 === screenshot2) {
            debugDevice('Screenshots are identical');
            return true;
        }
        const len1 = screenshot1.length;
        const len2 = screenshot2.length;
        debugDevice(`Screenshots differ: length1=${len1}, length2=${len2}`);
        if (Math.abs(len1 - len2) > 0.1 * Math.min(len1, len2)) {
            debugDevice('Screenshots have significant length difference');
            return false;
        }
        if (len1 > 0 && len2 > 0) {
            const minLength = Math.min(len1, len2);
            const sampleSize = Math.min(2000, minLength);
            let diffCount = 0;
            for(let i = 0; i < sampleSize; i++)if (screenshot1[i] !== screenshot2[i]) diffCount++;
            const diffPercent = diffCount / sampleSize * 100;
            debugDevice(`Character differences: ${diffCount}/${sampleSize} (${diffPercent.toFixed(2)}%)`);
            const isSimilar = diffPercent <= tolerancePercent;
            if (isSimilar) debugDevice(`Screenshots are similar enough (${diffPercent.toFixed(2)}% <= ${tolerancePercent}%)`);
            return isSimilar;
        }
        return false;
    }
    async scrollUntilBoundary(direction, startPoint, maxUnchangedCount = 1) {
        const maxAttempts = 20;
        const { width, height } = await this.size();
        let start;
        if (startPoint) start = {
            x: Math.round(startPoint.left),
            y: Math.round(startPoint.top)
        };
        else switch(direction){
            case 'up':
                start = {
                    x: Math.round(width / 2),
                    y: Math.round(0.2 * height)
                };
                break;
            case 'down':
                start = {
                    x: Math.round(width / 2),
                    y: Math.round(0.8 * height)
                };
                break;
            case 'left':
                start = {
                    x: Math.round(0.8 * width),
                    y: Math.round(height / 2)
                };
                break;
            case 'right':
                start = {
                    x: Math.round(0.2 * width),
                    y: Math.round(height / 2)
                };
                break;
        }
        let lastScreenshot = null;
        let unchangedCount = 0;
        debugDevice(`Starting scroll to ${direction} with content detection`);
        for(let i = 0; i < maxAttempts; i++)try {
            debugDevice(`Scroll attempt ${i + 1}/${maxAttempts}`);
            await sleep(500);
            const currentScreenshot = await this.screenshotBase64();
            if (lastScreenshot && this.compareScreenshots(lastScreenshot, currentScreenshot, 10)) {
                unchangedCount++;
                debugDevice(`Screen content unchanged (${unchangedCount}/${maxUnchangedCount})`);
                if (unchangedCount >= maxUnchangedCount) {
                    debugDevice(`Reached ${direction}: screen content no longer changes`);
                    break;
                }
            } else {
                if (lastScreenshot) debugDevice(`Content changed, resetting counter (was ${unchangedCount})`);
                unchangedCount = 0;
            }
            if (i >= 15 && 0 === unchangedCount) {
                debugDevice(`Too many attempts with dynamic content, stopping scroll to ${direction}`);
                break;
            }
            lastScreenshot = currentScreenshot;
            const scrollDistance = Math.round('left' === direction || 'right' === direction ? 0.6 * width : 0.6 * height);
            debugDevice(`Performing scroll: ${direction}, distance: ${scrollDistance}`);
            switch(direction){
                case 'up':
                    await this.swipe(start.x, start.y, start.x, start.y + scrollDistance, 300);
                    break;
                case 'down':
                    await this.swipe(start.x, start.y, start.x, start.y - scrollDistance, 300);
                    break;
                case 'left':
                    await this.swipe(start.x, start.y, start.x + scrollDistance, start.y, 300);
                    break;
                case 'right':
                    await this.swipe(start.x, start.y, start.x - scrollDistance, start.y, 300);
                    break;
            }
            debugDevice('Waiting for scroll and inertia to complete...');
            await sleep(2000);
        } catch (error) {
            debugDevice(`Error during scroll attempt ${i + 1}: ${error}`);
            await sleep(300);
        }
        debugDevice(`Scroll to ${direction} completed after ${maxAttempts} attempts`);
    }
    async scrollUntilLeft(startPoint) {
        await this.scrollUntilBoundary('left', startPoint, 1);
    }
    async scrollUntilRight(startPoint) {
        await this.scrollUntilBoundary('right', startPoint, 3);
    }
    async home() {
        await this.wdaBackend.pressHomeButton();
    }
    async appSwitcher() {
        try {
            debugDevice('Triggering app switcher with slow swipe up gesture');
            const { width, height } = await this.size();
            const centerX = Math.round(width / 2);
            const startY = Math.round(height - 5);
            const endY = Math.round(0.5 * height);
            await this.wdaBackend.swipe(centerX, startY, centerX, endY, 1500);
            await sleep(800);
        } catch (error) {
            debugDevice(`App switcher failed: ${error}`);
            throw new Error(`Failed to trigger app switcher: ${error}`);
        }
    }
    async hideKeyboard(keyNames) {
        try {
            const dismissKeys = keyNames && keyNames.length > 0 ? keyNames : [
                'return',
                'done',
                'go',
                'search',
                'next',
                'send'
            ];
            debugDevice(`Attempting to dismiss keyboard using WDA API with keys: ${dismissKeys.join(', ')}`);
            try {
                await this.wdaBackend.dismissKeyboard(dismissKeys);
                debugDevice('Successfully dismissed keyboard using WDA API');
                await sleep(500);
                return true;
            } catch (wdaError) {
                debugDevice(`WDA dismissKeyboard failed, falling back to swipe gesture: ${wdaError}`);
            }
            const windowSize = await this.wdaBackend.getWindowSize();
            const centerX = Math.round(windowSize.width / 2);
            const startY = Math.round(0.9 * windowSize.height);
            const endY = Math.round(0.5 * windowSize.height);
            await this.swipe(centerX, startY, centerX, endY, 300);
            debugDevice('Dismissed keyboard with swipe up gesture from bottom of screen');
            await sleep(500);
            return true;
        } catch (error) {
            debugDevice(`Failed to hide keyboard: ${error}`);
            return false;
        }
    }
    async openUrl(url, options) {
        const opts = {
            useSafariAsBackup: true,
            waitTime: 2000,
            ...options
        };
        try {
            debugDevice(`Opening URL: ${url}`);
            await this.wdaBackend.openUrl(url);
            await sleep(opts.waitTime);
            debugDevice(`Successfully opened URL: ${url}`);
        } catch (error) {
            debugDevice(`Direct URL opening failed: ${error}`);
            if (opts.useSafariAsBackup) {
                debugDevice(`Attempting to open URL via Safari: ${url}`);
                await this.openUrlViaSafari(url);
            } else throw new Error(`Failed to open URL: ${error}`);
        }
    }
    async openUrlViaSafari(url) {
        try {
            debugDevice(`Opening URL via Safari: ${url}`);
            await this.wdaBackend.terminateApp('com.apple.mobilesafari');
            await this.wdaBackend.launchApp('com.apple.mobilesafari');
            await sleep(2000);
            await this.typeText(url);
            await sleep(500);
            await this.pressKey('Return');
            await sleep(1000);
            try {
                await sleep(2000);
                debugDevice(`URL opened via Safari: ${url}`);
            } catch (dialogError) {
                debugDevice(`No confirmation dialog or dialog handling failed: ${dialogError}`);
            }
        } catch (error) {
            debugDevice(`Failed to open URL via Safari: ${error}`);
            throw new Error(`Failed to open URL via Safari: ${error}`);
        }
    }
    async runWdaRequest(method, endpoint, data) {
        return await this.wdaBackend.executeRequest(method, endpoint, data);
    }
    async destroy() {
        if (this.destroyed) return;
        try {
            await this.wdaBackend.deleteSession();
            await this.wdaManager.stop();
        } catch (error) {
            debugDevice(`Error during cleanup: ${error}`);
        }
        this.destroyed = true;
        debugDevice(`iOS device ${this.deviceId} destroyed`);
    }
    constructor(options){
        _define_property(this, "deviceId", void 0);
        _define_property(this, "devicePixelRatio", 1);
        _define_property(this, "devicePixelRatioInitialized", false);
        _define_property(this, "destroyed", false);
        _define_property(this, "description", void 0);
        _define_property(this, "customActions", void 0);
        _define_property(this, "wdaBackend", void 0);
        _define_property(this, "wdaManager", void 0);
        _define_property(this, "mjpegStreamUrl", void 0);
        _define_property(this, "appNameMapping", {});
        _define_property(this, "interfaceType", 'ios');
        _define_property(this, "uri", void 0);
        _define_property(this, "options", void 0);
        this.deviceId = 'pending-connection';
        this.options = options;
        this.customActions = options?.customActions;
        const wdaPort = options?.wdaPort || DEFAULT_WDA_PORT;
        const wdaHost = options?.wdaHost || 'localhost';
        const mjpegPort = options?.wdaMjpegPort ?? DEFAULT_WDA_MJPEG_PORT;
        this.wdaBackend = new IOSWebDriverClient({
            port: wdaPort,
            host: wdaHost
        });
        this.wdaManager = WDAManager.getInstance(wdaPort, wdaHost);
        this.mjpegStreamUrl = `http://${wdaHost}:${mjpegPort}`;
    }
}
const runWdaRequestParamSchema = z.object({
    method: z["enum"](WDA_HTTP_METHODS).describe('HTTP method (GET, POST, DELETE, PUT)'),
    endpoint: z.string().describe('WebDriver API endpoint'),
    data: z.object({}).passthrough().optional().describe('Optional request body data as JSON object')
});
const launchParamSchema = z.object({
    uri: z.string().describe('App name, bundle ID, or URL to launch. Prioritize using the exact bundle ID or URL the user has provided. If none provided, use the accurate app name.')
});
const terminateParamSchema = z.object({
    uri: z.string().describe('Bundle ID of the app to terminate (close). Use the exact bundle ID, e.g. com.apple.Preferences.')
});
const createPlatformActions = (device)=>({
        RunWdaRequest: defineAction({
            name: 'RunWdaRequest',
            description: 'Execute WebDriverAgent API request directly on iOS device',
            interfaceAlias: 'runWdaRequest',
            paramSchema: runWdaRequestParamSchema,
            sample: {
                method: 'GET',
                endpoint: '/status'
            },
            call: async (param)=>await device.runWdaRequest(param.method, param.endpoint, param.data)
        }),
        Launch: defineAction({
            name: 'Launch',
            description: 'Launch an iOS app or URL',
            interfaceAlias: 'launch',
            paramSchema: launchParamSchema,
            sample: {
                uri: 'com.apple.Preferences'
            },
            call: async (param)=>{
                if (!param.uri || '' === param.uri.trim()) throw new Error('Launch requires a non-empty uri parameter');
                await device.launch(param.uri);
            }
        }),
        Terminate: defineAction({
            name: 'Terminate',
            description: 'Terminate (close) an iOS app by its bundle ID',
            interfaceAlias: 'terminate',
            paramSchema: terminateParamSchema,
            sample: {
                uri: 'com.apple.Preferences'
            },
            delayBeforeRunner: 0,
            delayAfterRunner: 0,
            call: async (param)=>{
                if (!param.uri || '' === param.uri.trim()) throw new Error('Terminate requires a non-empty uri parameter');
                await device.terminate(param.uri);
            }
        }),
        IOSHomeButton: defineAction({
            name: 'IOSHomeButton',
            description: 'Trigger the system "home" operation on iOS devices',
            delayBeforeRunner: 0,
            delayAfterRunner: 0,
            call: async ()=>{
                await device.home();
            }
        }),
        IOSAppSwitcher: defineAction({
            name: 'IOSAppSwitcher',
            description: 'Trigger the system "app switcher" operation on iOS devices',
            call: async ()=>{
                await device.appSwitcher();
            }
        })
    });
function agent_define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
const debugAgent = getDebug('ios:agent');
class IOSAgent extends Agent {
    async launch(uri) {
        const action = this.wrapActionInActionSpace('Launch');
        return action({
            uri
        });
    }
    async terminate(uri) {
        const action = this.wrapActionInActionSpace('Terminate');
        return action({
            uri
        });
    }
    createActionWrapper(name) {
        const action = this.wrapActionInActionSpace(name);
        return (...args)=>action(args[0]);
    }
    constructor(device, opts){
        super(device, opts), agent_define_property(this, "runWdaRequest", void 0), agent_define_property(this, "home", void 0), agent_define_property(this, "appSwitcher", void 0), agent_define_property(this, "appNameMapping", void 0);
        this.appNameMapping = mergeAndNormalizeAppNameMapping(defaultAppNameMapping, opts?.appNameMapping);
        device.setAppNameMapping(this.appNameMapping);
        this.runWdaRequest = this.createActionWrapper('RunWdaRequest');
        this.home = this.createActionWrapper('IOSHomeButton');
        this.appSwitcher = this.createActionWrapper('IOSAppSwitcher');
    }
}
async function agentFromWebDriverAgent(opts) {
    debugAgent('Creating iOS agent with WebDriverAgent');
    const overrideModule = opts?.iOSDeviceClassOverride?.trim() || process.env[MIDSCENE_IOS_DEVICE_CLASS_OVERRIDE]?.trim();
    let DeviceClass = IOSDevice;
    if (overrideModule) try {
        const overrideExports = await import(overrideModule);
        const overrideDeviceClass = Object.prototype.hasOwnProperty.call(overrideExports, 'IOSDevice') ? overrideExports.IOSDevice : overrideExports.default;
        if ('function' != typeof overrideDeviceClass) throw new Error(`Module "${overrideModule}" does not export a valid iOS device class (expected "IOSDevice" or default export).`);
        DeviceClass = overrideDeviceClass;
    } catch (error) {
        throw new Error(`Failed to load iOS device class override from "${overrideModule}". Please make sure the package is installed and exports IOSDevice (or default) with Midscene-compatible methods. Original error: ${error instanceof Error ? error.message : String(error)}`);
    }
    const device = new DeviceClass(opts || {});
    await device.connect();
    return new IOSAgent(device, opts);
}
function mcp_tools_define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
const debug = getDebug('mcp:ios-tools');
const iosInitArgShape = {
    deviceId: z.string().optional().describe('iOS device UDID (optional when WDA auto-detect is sufficient)'),
    wdaHost: z.string().optional().describe('WebDriverAgent host, defaults to localhost'),
    wdaPort: z.number().optional().describe('WebDriverAgent port'),
    useWDA: z.boolean().optional().describe('Whether to reuse an existing WebDriverAgent session'),
    wdaMjpegPort: z.number().optional().describe('WebDriverAgent MJPEG streaming port')
};
class IOSMidsceneTools extends BaseMidsceneTools {
    getCliReportSessionName() {
        return 'midscene-ios';
    }
    createTemporaryDevice() {
        return new IOSDevice({});
    }
    async ensureAgent(opts) {
        const hasOpts = !!opts && Object.keys(opts).length > 0;
        const nextSignature = hasOpts ? JSON.stringify(opts) : void 0;
        if (this.agent && hasOpts && nextSignature !== this.lastOptsSignature) {
            try {
                await this.agent.destroy?.();
            } catch (error) {
                debug('Failed to destroy agent during cleanup:', error);
            }
            this.agent = void 0;
        }
        if (this.agent) return this.agent;
        debug('Creating iOS agent with WebDriverAgent options:', opts || {});
        const reportOptions = this.readCliReportAgentOptions();
        this.agent = await agentFromWebDriverAgent({
            autoDismissKeyboard: false,
            ...reportOptions ?? {},
            ...opts ?? {}
        });
        this.lastOptsSignature = nextSignature;
        return this.agent;
    }
    preparePlatformTools() {
        return [
            {
                name: 'ios_connect',
                description: 'Connect to iOS device or simulator via WebDriverAgent',
                schema: this.getAgentInitArgSchema(),
                cli: this.getAgentInitArgCliMetadata(),
                handler: async (args)=>{
                    const initArgs = this.extractAgentInitParam(args);
                    const identity = initArgs?.deviceId ?? (initArgs?.wdaHost || initArgs?.wdaPort ? `wda-${initArgs.wdaHost ?? 'localhost'}-${initArgs.wdaPort ?? 'default'}` : 'auto');
                    const reportSession = this.createNewCliReportSession(identity);
                    this.commitCliReportSession(reportSession);
                    if (this.agent) {
                        try {
                            await this.agent.destroy?.();
                        } catch (error) {
                            debug('Failed to destroy agent during connect:', error);
                        }
                        this.agent = void 0;
                    }
                    const agent = await this.ensureAgent(initArgs);
                    const screenshot = await agent.page.screenshotBase64();
                    return {
                        content: [
                            {
                                type: 'text',
                                text: `Connected to iOS device${initArgs?.deviceId ? `: ${initArgs.deviceId}` : ''}`
                            },
                            ...this.buildScreenshotContent(screenshot)
                        ],
                        isError: false
                    };
                }
            },
            {
                name: 'ios_disconnect',
                description: 'Disconnect from current iOS device and release WebDriverAgent resources',
                schema: {},
                handler: this.createDisconnectHandler('iOS device')
            }
        ];
    }
    constructor(...args){
        super(...args), mcp_tools_define_property(this, "initArgSpec", {
            namespace: 'ios',
            shape: iosInitArgShape,
            cli: {
                preferBareKeys: true
            },
            adapt: (extracted)=>extracted
        }), mcp_tools_define_property(this, "lastOptsSignature", void 0);
    }
}
const tools = new IOSMidsceneTools();
runToolsCLI(tools, 'midscene-ios', {
    stripPrefix: 'ios_',
    version: "1.7.9",
    extraCommands: createReportCliCommands()
}).catch((e)=>{
    process.exit(reportCLIError(e));
});
