// @midscene/computer/dist/es/index.mjs
import node_assert from "node:assert";
import { execFileSync, execSync, spawn, spawnSync } from "node:child_process";
import { existsSync } from "node:fs";
import { createRequire } from "node:module";
import { dirname, resolve as external_node_path_resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { getMidsceneLocationSchema, z } from "@midscene/core";
import { actionHoverParamSchema, actionTapParamSchema, defineAction, defineActionClearInput, defineActionDoubleClick, defineActionDragAndDrop, defineActionHover, defineActionInput, defineActionKeyboardPress, defineActionRightClick, defineActionScroll, defineActionTap } from "@midscene/core/device";
import { sleep } from "@midscene/core/utils";
import { createImgBase64ByFormat } from "@midscene/shared/img";
import { getDebug } from "@midscene/shared/logger";
import screenshot_desktop from "screenshot-desktop";
import { Agent } from "@midscene/core/agent";
import { once } from "node:events";
import { createInterface } from "node:readline";
import { BaseMidsceneTools } from "@midscene/shared/mcp/base-tools";
import { overrideAIConfig } from "@midscene/shared/env";
const debugXvfb = getDebug('computer:xvfb');
function checkXvfbInstalled() {
    try {
        execSync('which Xvfb', {
            stdio: 'ignore'
        });
        return true;
    } catch  {
        return false;
    }
}
function findAvailableDisplay(startFrom = 99) {
    for(let n = startFrom; n < startFrom + 100; n++)if (!existsSync(`/tmp/.X${n}-lock`)) return n;
    throw new Error(`No available display number found (checked ${startFrom} to ${startFrom + 99})`);
}
function startXvfb(options) {
    const resolution = options?.resolution || '1920x1080x24';
    const displayNum = options?.displayNumber ?? findAvailableDisplay();
    const display = `:${displayNum}`;
    return new Promise((resolve, reject)=>{
        debugXvfb(`Starting Xvfb on display ${display} with resolution ${resolution}`);
        const xvfbProcess = spawn('Xvfb', [
            display,
            '-screen',
            '0',
            resolution,
            '-ac',
            '-nolisten',
            'tcp'
        ], {
            stdio: 'ignore'
        });
        let settled = false;
        xvfbProcess.on('error', (err)=>{
            if (!settled) {
                settled = true;
                reject(new Error(`Failed to start Xvfb: ${err.message}`));
            }
        });
        xvfbProcess.on('exit', (code)=>{
            if (!settled) {
                settled = true;
                reject(new Error(`Xvfb exited unexpectedly with code ${code}`));
            }
        });
        const instance = {
            process: xvfbProcess,
            display,
            stop () {
                try {
                    xvfbProcess.kill('SIGTERM');
                } catch  {}
            }
        };
        setTimeout(()=>{
            if (!settled) {
                settled = true;
                debugXvfb(`Xvfb started on display ${display}`);
                resolve(instance);
            }
        }, 500);
    });
}
function needsXvfb(explicitOpt) {
    if ('linux' !== process.platform) return false;
    return true === explicitOpt;
}
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
const computerInputParamSchema = z.object({
    value: z.string().describe('The text to input'),
    mode: z["enum"]([
        'replace',
        'clear',
        'append'
    ]).default('replace').optional().describe('Input mode: replace, clear, or append'),
    locate: getMidsceneLocationSchema().describe('The input field to be filled').optional()
});
const SMOOTH_MOVE_STEPS_TAP = 8;
const SMOOTH_MOVE_STEPS_MOUSE_MOVE = 10;
const SMOOTH_MOVE_DELAY_TAP = 8;
const SMOOTH_MOVE_DELAY_MOUSE_MOVE = 10;
const MOUSE_MOVE_EFFECT_WAIT = 300;
const CLICK_HOLD_DURATION = 50;
const INPUT_FOCUS_DELAY = 300;
const INPUT_CLEAR_DELAY = 150;
const SCROLL_REPEAT_COUNT = 10;
const SCROLL_STEP_DELAY = 100;
const SCROLL_COMPLETE_DELAY = 500;
const EDGE_SCROLL_TOTAL_PX = 50000;
const EDGE_SCROLL_STEPS = 400;
const PHASED_PIXELS_PER_STEP = 30;
const PHASED_MIN_STEPS = 10;
const APPROX_VIEWPORT_HEIGHT_PX = 600;
const EDGE_SCROLL_SPEC = {
    scrollToTop: {
        direction: 'up',
        key: 'home',
        libnut: [
            0,
            10
        ]
    },
    scrollToBottom: {
        direction: 'down',
        key: 'end',
        libnut: [
            0,
            -10
        ]
    },
    scrollToLeft: {
        direction: 'left',
        key: 'home',
        libnut: [
            -10,
            0
        ]
    },
    scrollToRight: {
        direction: 'right',
        key: 'end',
        libnut: [
            10,
            0
        ]
    }
};
const APPLESCRIPT_KEY_CODE_MAP = {
    return: 36,
    enter: 36,
    tab: 48,
    space: 49,
    backspace: 51,
    delete: 51,
    escape: 53,
    forwarddelete: 117,
    left: 123,
    right: 124,
    down: 125,
    up: 126,
    home: 115,
    end: 119,
    pageup: 116,
    pagedown: 121,
    f1: 122,
    f2: 120,
    f3: 99,
    f4: 118,
    f5: 96,
    f6: 97,
    f7: 98,
    f8: 100,
    f9: 101,
    f10: 109,
    f11: 103,
    f12: 111
};
const APPLESCRIPT_MODIFIER_MAP = {
    command: 'command down',
    cmd: 'command down',
    control: 'control down',
    ctrl: 'control down',
    shift: 'shift down',
    alt: 'option down',
    option: 'option down',
    meta: 'command down'
};
function sendKeyViaAppleScript(key, modifiers = []) {
    const lowerKey = key.toLowerCase();
    const keyCode = APPLESCRIPT_KEY_CODE_MAP[lowerKey];
    const modifierParts = modifiers.map((m)=>APPLESCRIPT_MODIFIER_MAP[m.toLowerCase()]).filter(Boolean);
    const modifierStr = modifierParts.length > 0 ? ` using {${modifierParts.join(', ')}}` : '';
    let script;
    if (void 0 !== keyCode) script = `tell application "System Events" to key code ${keyCode}${modifierStr}`;
    else {
        const escapedKey = key.replace(/\\/g, '\\\\').replace(/"/g, '\\"');
        script = `tell application "System Events" to keystroke "${escapedKey}"${modifierStr}`;
    }
    debugDevice('sendKeyViaAppleScript', {
        key,
        modifiers,
        script
    });
    execFileSync("osascript", [
        '-e',
        script
    ]);
}
let device_libnut = null;
let libnutLoadError = null;
async function getLibnut() {
    if (device_libnut) return device_libnut;
    if (libnutLoadError) throw libnutLoadError;
    try {
        const require = createRequire(import.meta.url);
        const libnutModule = require('@computer-use/libnut/dist/import_libnut');
        device_libnut = libnutModule.libnut;
        if (!device_libnut) throw new Error('libnut module loaded but libnut object is undefined');
        return device_libnut;
    } catch (error) {
        libnutLoadError = error;
        throw new Error(`Failed to load @computer-use/libnut. Make sure it is properly installed and compiled for your platform. Error: ${error}`);
    }
}
const debugDevice = getDebug('computer:device');
let phasedScrollBinaryPath;
function getPhasedScrollBinary() {
    if (void 0 !== phasedScrollBinaryPath) return phasedScrollBinaryPath;
    if ('darwin' !== process.platform) {
        phasedScrollBinaryPath = null;
        return null;
    }
    const require = createRequire(import.meta.url);
    let pkgRoot = null;
    try {
        pkgRoot = dirname(require.resolve('@midscene/computer/package.json'));
    } catch  {
        const hereDir = dirname(fileURLToPath(import.meta.url));
        for (const candidate of [
            external_node_path_resolve(hereDir, '..'),
            external_node_path_resolve(hereDir, '../..')
        ])if (existsSync(external_node_path_resolve(candidate, 'package.json'))) {
            pkgRoot = candidate;
            break;
        }
    }
    if (!pkgRoot) {
        debugDevice('phased-scroll: cannot locate @midscene/computer package root');
        phasedScrollBinaryPath = null;
        return null;
    }
    const binPath = external_node_path_resolve(pkgRoot, 'bin/darwin/phased-scroll');
    if (!existsSync(binPath)) {
        debugDevice('phased-scroll binary not found at', binPath);
        phasedScrollBinaryPath = null;
        return null;
    }
    phasedScrollBinaryPath = binPath;
    return binPath;
}
let phasedScrollExecWarned = false;
function runPhasedScroll(direction, pixels, steps) {
    const bin = getPhasedScrollBinary();
    if (!bin) return false;
    try {
        const res = spawnSync(bin, [
            direction,
            String(Math.max(1, Math.round(pixels))),
            String(steps)
        ], {
            stdio: 'ignore'
        });
        if (0 === res.status) return true;
        if (!phasedScrollExecWarned) {
            phasedScrollExecWarned = true;
            console.warn(`[@midscene/computer] phased-scroll helper exited with status ${res.status}; falling back to keyboard/libnut. This usually means Accessibility permission has not been granted to the host process.`);
        }
        debugDevice('phased-scroll exited non-zero', res.status, res.error);
        return false;
    } catch (err) {
        if (!phasedScrollExecWarned) {
            phasedScrollExecWarned = true;
            console.warn(`[@midscene/computer] phased-scroll helper failed to spawn (${err?.message}); falling back to keyboard/libnut.`);
        }
        debugDevice('phased-scroll spawn failed', err);
        return false;
    }
}
async function smoothMoveMouse(targetX, targetY, steps, stepDelay) {
    node_assert(device_libnut, 'libnut not initialized');
    const currentPos = device_libnut.getMousePos();
    for(let i = 1; i <= steps; i++){
        const stepX = Math.round(currentPos.x + (targetX - currentPos.x) * i / steps);
        const stepY = Math.round(currentPos.y + (targetY - currentPos.y) * i / steps);
        device_libnut.moveMouse(stepX, stepY);
        await sleep(stepDelay);
    }
}
const KEY_NAME_MAP = {
    windows: 'win',
    win: 'win',
    ctrl: 'control',
    esc: 'escape',
    del: 'delete',
    ins: 'insert',
    pgup: 'pageup',
    pgdn: 'pagedown',
    arrowup: 'up',
    arrowdown: 'down',
    arrowleft: 'left',
    arrowright: 'right',
    volumedown: 'audio_vol_down',
    volumeup: 'audio_vol_up',
    mediavolumedown: 'audio_vol_down',
    mediavolumeup: 'audio_vol_up',
    mute: 'audio_mute',
    mediamute: 'audio_mute',
    mediaplay: 'audio_play',
    mediapause: 'audio_pause',
    mediaplaypause: 'audio_play',
    mediastop: 'audio_stop',
    medianexttrack: 'audio_next',
    mediaprevioustrack: 'audio_prev',
    medianext: 'audio_next',
    mediaprev: 'audio_prev'
};
const PRIMARY_KEY_MAP = {
    command: 'cmd',
    cmd: 'cmd',
    meta: 'meta',
    control: 'control',
    ctrl: 'control',
    shift: 'shift',
    alt: 'alt',
    option: 'alt'
};
function normalizeKeyName(key) {
    const lowerKey = key.toLowerCase();
    return KEY_NAME_MAP[lowerKey] || lowerKey;
}
function normalizePrimaryKey(key) {
    const lowerKey = key.toLowerCase();
    if (PRIMARY_KEY_MAP[lowerKey]) return PRIMARY_KEY_MAP[lowerKey];
    return KEY_NAME_MAP[lowerKey] || lowerKey;
}
class ComputerDevice {
    describe() {
        return this.description || 'Computer Device';
    }
    static async listDisplays() {
        try {
            const displays = await screenshot_desktop.listDisplays();
            return displays.map((d)=>({
                    id: String(d.id),
                    name: d.name || `Display ${d.id}`,
                    primary: d.primary || false
                }));
        } catch (error) {
            debugDevice(`Failed to list displays: ${error}`);
            return [];
        }
    }
    async connect() {
        debugDevice('Connecting to computer device');
        try {
            const headless = this.options?.headless ?? 'true' === process.env.MIDSCENE_COMPUTER_HEADLESS_LINUX;
            if (needsXvfb(headless)) {
                if (!checkXvfbInstalled()) throw new Error('Xvfb is required for headless mode but not installed. Install: sudo apt-get install xvfb');
                this.xvfbInstance = await startXvfb({
                    resolution: this.options?.xvfbResolution
                });
                process.env.DISPLAY = this.xvfbInstance.display;
                debugDevice(`Xvfb started on display ${this.xvfbInstance.display}`);
                this.xvfbCleanup = ()=>{
                    if (this.xvfbInstance) {
                        this.xvfbInstance.stop();
                        this.xvfbInstance = void 0;
                    }
                };
                process.on('exit', this.xvfbCleanup);
                process.on('SIGINT', this.xvfbCleanup);
                process.on('SIGTERM', this.xvfbCleanup);
            }
            device_libnut = await getLibnut();
            const size = await this.size();
            const displays = await ComputerDevice.listDisplays();
            const headlessInfo = this.xvfbInstance ? `\nHeadless: true (Xvfb on ${this.xvfbInstance.display})` : '';
            this.description = `
Type: Computer
Platform: ${process.platform}
Display: ${this.displayId || 'Primary'}
Screen Size: ${size.width}x${size.height}
Available Displays: ${displays.length > 0 ? displays.map((d)=>d.name).join(', ') : 'Unknown'}${headlessInfo}
`;
            debugDevice('Computer device connected', this.description);
            await this.healthCheck();
        } catch (error) {
            if (this.xvfbInstance) {
                this.xvfbInstance.stop();
                this.xvfbInstance = void 0;
            }
            debugDevice(`Failed to connect: ${error}`);
            throw new Error(`Unable to connect to computer device: ${error}`);
        }
    }
    async healthCheck() {
        console.log('[HealthCheck] Starting health check...');
        console.log("[HealthCheck] @midscene/computer v1.7.9");
        console.log('[HealthCheck] Taking screenshot...');
        const screenshotTimeout = 15000;
        let timeoutId;
        const timeoutPromise = new Promise((_, reject)=>{
            timeoutId = setTimeout(()=>reject(new Error('Screenshot timed out')), screenshotTimeout);
        });
        const base64 = await Promise.race([
            this.screenshotBase64().finally(()=>clearTimeout(timeoutId)),
            timeoutPromise
        ]);
        console.log(`[HealthCheck] Screenshot succeeded (length=${base64.length})`);
        console.log('[HealthCheck] Moving mouse...');
        node_assert(device_libnut, 'libnut not initialized');
        const startPos = device_libnut.getMousePos();
        console.log(`[HealthCheck] Current mouse position: (${startPos.x}, ${startPos.y})`);
        const offsetX = Math.floor(40 * Math.random()) + 10;
        const offsetY = Math.floor(40 * Math.random()) + 10;
        const targetX = startPos.x + offsetX;
        const targetY = startPos.y + offsetY;
        console.log(`[HealthCheck] Moving mouse to (${targetX}, ${targetY})...`);
        device_libnut.moveMouse(targetX, targetY);
        await sleep(50);
        const movedPos = device_libnut.getMousePos();
        console.log(`[HealthCheck] Mouse position after move: (${movedPos.x}, ${movedPos.y})`);
        const deltaX = Math.abs(movedPos.x - targetX);
        const deltaY = Math.abs(movedPos.y - targetY);
        if (deltaX > 5 || deltaY > 5) {
            const msg = `[HealthCheck] WARNING: Mouse control may not be working. Expected (${targetX}, ${targetY}), got (${movedPos.x}, ${movedPos.y}), delta=(${deltaX}, ${deltaY})`;
            console.warn(msg);
            debugDevice(msg);
            if ('win32' === process.platform && !this.isRunningAsAdmin()) {
                const hint = "Midscene is NOT running as Administrator. Windows blocks mouse/keyboard input to elevated (admin) applications from non-admin processes (UIPI). Please run your terminal or Node.js as Administrator and try again.";
                console.error(`\n[HealthCheck] ${hint}\n`);
                debugDevice(hint);
            }
        }
        device_libnut.moveMouse(startPos.x, startPos.y);
        console.log(`[HealthCheck] Mouse restored to (${startPos.x}, ${startPos.y})`);
        console.log('[HealthCheck] Listing monitors...');
        const displays = await ComputerDevice.listDisplays();
        if (displays.length > 0) {
            console.log(`[HealthCheck] Found ${displays.length} monitor(s):`);
            for (const display of displays){
                const primaryTag = display.primary ? ' (primary)' : '';
                console.log(`[HealthCheck]   - id=${display.id}, name=${display.name}${primaryTag}`);
            }
        } else console.log('[HealthCheck] No monitors detected');
        console.log('[HealthCheck] Health check passed');
    }
    isRunningAsAdmin() {
        if ('win32' !== process.platform) return false;
        try {
            execSync('net session', {
                stdio: 'pipe'
            });
            return true;
        } catch  {
            return false;
        }
    }
    async screenshotBase64() {
        if (this.destroyed) throw new Error('ComputerDevice has been destroyed');
        debugDevice('Taking screenshot', {
            displayId: this.displayId
        });
        try {
            const options = {
                format: 'png'
            };
            if (void 0 !== this.displayId) if ('darwin' === process.platform) {
                const screenIndex = Number(this.displayId);
                if (!Number.isNaN(screenIndex)) options.screen = screenIndex;
            } else options.screen = this.displayId;
            debugDevice('Screenshot options', options);
            const buffer = await screenshot_desktop(options);
            return createImgBase64ByFormat('png', buffer.toString('base64'));
        } catch (error) {
            debugDevice(`Screenshot failed: ${error}`);
            throw new Error(`Failed to take screenshot: ${error}`);
        }
    }
    async size() {
        node_assert(device_libnut, 'libnut not initialized');
        try {
            const screenSize = device_libnut.getScreenSize();
            return {
                width: screenSize.width,
                height: screenSize.height
            };
        } catch (error) {
            debugDevice(`Failed to get screen size: ${error}`);
            throw new Error(`Failed to get screen size: ${error}`);
        }
    }
    async typeViaClipboard(text) {
        node_assert(device_libnut, 'libnut not initialized');
        debugDevice('Using clipboard to input text', {
            textLength: text.length,
            preview: text.substring(0, 20)
        });
        const clipboardy = await import("clipboardy");
        const oldClipboard = await clipboardy.default.read().catch(()=>'');
        try {
            await clipboardy.default.write(text);
            await sleep(50);
            if (this.useAppleScript) sendKeyViaAppleScript('v', [
                'command'
            ]);
            else {
                const modifier = 'darwin' === process.platform ? 'command' : 'control';
                device_libnut.keyTap('v', [
                    modifier
                ]);
            }
            await sleep(100);
        } finally{
            if (oldClipboard) await clipboardy.default.write(oldClipboard).catch(()=>{
                debugDevice('Failed to restore clipboard content');
            });
        }
    }
    async smartTypeString(text) {
        node_assert(device_libnut, 'libnut not initialized');
        await this.typeViaClipboard(text);
    }
    actionSpace() {
        const defaultActions = [
            defineActionTap(async (param)=>{
                node_assert(device_libnut, 'libnut not initialized');
                const element = param.locate;
                node_assert(element, 'Element not found, cannot tap');
                const [x, y] = element.center;
                const targetX = Math.round(x);
                const targetY = Math.round(y);
                await smoothMoveMouse(targetX, targetY, SMOOTH_MOVE_STEPS_TAP, SMOOTH_MOVE_DELAY_TAP);
                device_libnut.mouseToggle('down', 'left');
                await sleep(CLICK_HOLD_DURATION);
                device_libnut.mouseToggle('up', 'left');
            }),
            defineActionDoubleClick(async (param)=>{
                node_assert(device_libnut, 'libnut not initialized');
                const element = param.locate;
                node_assert(element, 'Element not found, cannot double click');
                const [x, y] = element.center;
                device_libnut.moveMouse(Math.round(x), Math.round(y));
                device_libnut.mouseClick('left', true);
            }),
            defineActionRightClick(async (param)=>{
                node_assert(device_libnut, 'libnut not initialized');
                const element = param.locate;
                node_assert(element, 'Element not found, cannot right click');
                const [x, y] = element.center;
                device_libnut.moveMouse(Math.round(x), Math.round(y));
                device_libnut.mouseClick('right');
            }),
            defineAction({
                name: 'MouseMove',
                description: 'Move the mouse to the element',
                interfaceAlias: 'aiHover',
                paramSchema: actionHoverParamSchema,
                sample: {
                    locate: {
                        prompt: 'the navigation menu item "Products"'
                    }
                },
                call: async (param)=>{
                    node_assert(device_libnut, 'libnut not initialized');
                    const element = param.locate;
                    node_assert(element, 'Element not found, cannot move mouse');
                    const [x, y] = element.center;
                    const targetX = Math.round(x);
                    const targetY = Math.round(y);
                    await smoothMoveMouse(targetX, targetY, SMOOTH_MOVE_STEPS_MOUSE_MOVE, SMOOTH_MOVE_DELAY_MOUSE_MOVE);
                    await sleep(MOUSE_MOVE_EFFECT_WAIT);
                }
            }),
            defineAction({
                name: 'Input',
                description: 'Input text into the input field',
                interfaceAlias: 'aiInput',
                paramSchema: computerInputParamSchema,
                sample: {
                    value: 'test@example.com',
                    locate: {
                        prompt: 'the email input field'
                    }
                },
                call: async (param)=>{
                    node_assert(device_libnut, 'libnut not initialized');
                    const element = param.locate;
                    if (element) {
                        const [x, y] = element.center;
                        device_libnut.moveMouse(Math.round(x), Math.round(y));
                        device_libnut.mouseClick('left');
                        await sleep(INPUT_FOCUS_DELAY);
                        if ('append' !== param.mode) {
                            if (this.useAppleScript) {
                                sendKeyViaAppleScript('a', [
                                    'command'
                                ]);
                                await sleep(50);
                                sendKeyViaAppleScript('backspace', []);
                            } else {
                                const modifier = 'darwin' === process.platform ? 'command' : 'control';
                                device_libnut.keyTap('a', [
                                    modifier
                                ]);
                                await sleep(50);
                                device_libnut.keyTap('backspace');
                            }
                            await sleep(INPUT_CLEAR_DELAY);
                        }
                    }
                    if ('clear' === param.mode) return;
                    if (!param.value) return;
                    await this.smartTypeString(param.value);
                }
            }),
            defineActionScroll(async (param)=>{
                node_assert(device_libnut, 'libnut not initialized');
                if (param.locate) {
                    const element = param.locate;
                    const [x, y] = element.center;
                    device_libnut.moveMouse(Math.round(x), Math.round(y));
                }
                const scrollType = param?.scrollType;
                const edgeSpec = scrollType && scrollType in EDGE_SCROLL_SPEC ? EDGE_SCROLL_SPEC[scrollType] : null;
                if (edgeSpec) {
                    if (runPhasedScroll(edgeSpec.direction, EDGE_SCROLL_TOTAL_PX, EDGE_SCROLL_STEPS)) return void await sleep(SCROLL_COMPLETE_DELAY);
                    if (this.useAppleScript) {
                        sendKeyViaAppleScript(edgeSpec.key);
                        await sleep(SCROLL_COMPLETE_DELAY);
                        return;
                    }
                    const [dx, dy] = edgeSpec.libnut;
                    for(let i = 0; i < SCROLL_REPEAT_COUNT; i++){
                        device_libnut.scrollMouse(dx, dy);
                        await sleep(SCROLL_STEP_DELAY);
                    }
                    return;
                }
                if ('singleAction' === scrollType || !scrollType) {
                    const distance = param?.distance || 500;
                    const direction = param?.direction || 'down';
                    const isKnownDirection = 'up' === direction || 'down' === direction || 'left' === direction || 'right' === direction;
                    if (isKnownDirection) {
                        const steps = Math.max(PHASED_MIN_STEPS, Math.round(distance / PHASED_PIXELS_PER_STEP));
                        if (runPhasedScroll(direction, distance, steps)) return void await sleep(SCROLL_COMPLETE_DELAY);
                    }
                    if (this.useAppleScript && ('up' === direction || 'down' === direction)) {
                        const pages = Math.max(1, Math.round(distance / APPROX_VIEWPORT_HEIGHT_PX));
                        const key = 'up' === direction ? 'pageup' : 'pagedown';
                        for(let i = 0; i < pages; i++){
                            sendKeyViaAppleScript(key);
                            await sleep(SCROLL_STEP_DELAY);
                        }
                        await sleep(SCROLL_COMPLETE_DELAY);
                        return;
                    }
                    const ticks = Math.ceil(distance / 100);
                    const directionMap = {
                        up: [
                            0,
                            ticks
                        ],
                        down: [
                            0,
                            -ticks
                        ],
                        left: [
                            -ticks,
                            0
                        ],
                        right: [
                            ticks,
                            0
                        ]
                    };
                    const [dx, dy] = directionMap[direction] || [
                        0,
                        -ticks
                    ];
                    device_libnut.scrollMouse(dx, dy);
                    await sleep(SCROLL_COMPLETE_DELAY);
                    return;
                }
                throw new Error(`Unknown scroll type: ${scrollType}, param: ${JSON.stringify(param)}`);
            }),
            defineActionKeyboardPress(async (param)=>{
                node_assert(device_libnut, 'libnut not initialized');
                if (param.locate) {
                    const [x, y] = param.locate.center;
                    device_libnut.moveMouse(Math.round(x), Math.round(y));
                    device_libnut.mouseClick('left');
                    await sleep(50);
                }
                const keys = param.keyName.split('+');
                const modifiers = keys.slice(0, -1).map(normalizeKeyName);
                const key = normalizePrimaryKey(keys[keys.length - 1]);
                debugDevice('KeyboardPress', {
                    original: param.keyName,
                    key,
                    modifiers,
                    driver: this.useAppleScript ? "applescript" : 'libnut'
                });
                if (this.useAppleScript) sendKeyViaAppleScript(key, modifiers);
                else if (modifiers.length > 0) device_libnut.keyTap(key, modifiers);
                else device_libnut.keyTap(key);
            }),
            defineActionDragAndDrop(async (param)=>{
                node_assert(device_libnut, 'libnut not initialized');
                const from = param.from;
                const to = param.to;
                node_assert(from, 'missing "from" param for drag and drop');
                node_assert(to, 'missing "to" param for drag and drop');
                const [fromX, fromY] = from.center;
                const [toX, toY] = to.center;
                device_libnut.moveMouse(Math.round(fromX), Math.round(fromY));
                device_libnut.mouseToggle('down', 'left');
                await sleep(100);
                device_libnut.moveMouse(Math.round(toX), Math.round(toY));
                await sleep(100);
                device_libnut.mouseToggle('up', 'left');
            }),
            defineActionClearInput(async (param)=>{
                node_assert(device_libnut, 'libnut not initialized');
                const element = param.locate;
                node_assert(element, 'Element not found, cannot clear input');
                const [x, y] = element.center;
                device_libnut.moveMouse(Math.round(x), Math.round(y));
                device_libnut.mouseClick('left');
                await sleep(100);
                if (this.useAppleScript) {
                    sendKeyViaAppleScript('a', [
                        'command'
                    ]);
                    await sleep(50);
                    sendKeyViaAppleScript('backspace', []);
                } else {
                    const modifier = 'darwin' === process.platform ? 'command' : 'control';
                    device_libnut.keyTap('a', [
                        modifier
                    ]);
                    device_libnut.keyTap('backspace');
                }
                await sleep(50);
            })
        ];
        const platformActions = Object.values(createPlatformActions());
        const customActions = this.options?.customActions || [];
        return [
            ...defaultActions,
            ...platformActions,
            ...customActions
        ];
    }
    async destroy() {
        if (this.destroyed) return;
        if (this.xvfbInstance) {
            this.xvfbInstance.stop();
            this.xvfbInstance = void 0;
        }
        if (this.xvfbCleanup) {
            process.removeListener('exit', this.xvfbCleanup);
            process.removeListener('SIGINT', this.xvfbCleanup);
            process.removeListener('SIGTERM', this.xvfbCleanup);
            this.xvfbCleanup = void 0;
        }
        this.destroyed = true;
        debugDevice('Computer device destroyed');
    }
    async url() {
        return '';
    }
    constructor(options){
        _define_property(this, "interfaceType", 'computer');
        _define_property(this, "options", void 0);
        _define_property(this, "displayId", void 0);
        _define_property(this, "description", void 0);
        _define_property(this, "destroyed", false);
        _define_property(this, "xvfbInstance", void 0);
        _define_property(this, "xvfbCleanup", void 0);
        _define_property(this, "useAppleScript", void 0);
        _define_property(this, "uri", void 0);
        this.options = options;
        this.displayId = options?.displayId;
        this.useAppleScript = 'darwin' === process.platform && options?.keyboardDriver !== 'libnut';
    }
}
function createPlatformActions() {
    return {
        ListDisplays: defineAction({
            name: 'ListDisplays',
            description: 'List all available displays/monitors',
            call: async ()=>await ComputerDevice.listDisplays()
        })
    };
}
const platformBinaryMap = {
    darwin: {
        directory: 'darwin',
        fileName: 'rdp-helper'
    },
    linux: {
        directory: 'linux',
        fileName: 'rdp-helper'
    },
    win32: {
        directory: 'win32',
        fileName: 'rdp-helper.exe'
    }
};
function getPlatformBinary(platform) {
    if (platform in platformBinaryMap) return platformBinaryMap[platform];
}
function currentDirname() {
    if ('undefined' != typeof __dirname) return __dirname;
    return dirname(fileURLToPath(import.meta.url));
}
function getRdpHelperBinaryPath() {
    const platformBinary = getPlatformBinary(process.platform);
    if (!platformBinary) throw new Error(`@midscene/computer RDP helper does not support platform ${process.platform}`);
    const hereDir = currentDirname();
    const candidateRoots = [
        external_node_path_resolve(hereDir, '../..'),
        external_node_path_resolve(hereDir, '../../..')
    ];
    for (const root of candidateRoots){
        const binaryPath = external_node_path_resolve(root, 'bin', platformBinary.directory, platformBinary.fileName);
        if (existsSync(binaryPath)) return binaryPath;
    }
    throw new Error(`RDP helper binary not found for ${process.platform}. Run \`pnpm --filter @midscene/computer run build:native\` first.`);
}
function backend_client_define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
const debug = getDebug('rdp:backend');
const HELPER_SHUTDOWN_TIMEOUT_MS = 3000;
const MAX_STDERR_LINES = 40;
const notImplementedError = ()=>new Error('RDP backend transport is not implemented yet. Inject a custom backend into RDPDevice or implement a real helper transport.');
class UnsupportedRDPBackendClient {
    async connect(_config) {
        debug('connect called on unsupported backend');
        throw notImplementedError();
    }
    async disconnect() {
        debug('disconnect called on unsupported backend');
        throw notImplementedError();
    }
    async screenshotBase64() {
        debug('screenshotBase64 called on unsupported backend');
        throw notImplementedError();
    }
    async size() {
        debug('size called on unsupported backend');
        throw notImplementedError();
    }
    async mouseMove(_x, _y) {
        debug('mouseMove called on unsupported backend');
        throw notImplementedError();
    }
    async mouseButton(_button, _action) {
        debug('mouseButton called on unsupported backend');
        throw notImplementedError();
    }
    async wheel(_direction, _amount, _x, _y) {
        debug('wheel called on unsupported backend');
        throw notImplementedError();
    }
    async keyPress(_keyName) {
        debug('keyPress called on unsupported backend');
        throw notImplementedError();
    }
    async typeText(_text) {
        debug('typeText called on unsupported backend');
        throw notImplementedError();
    }
    async clearInput() {
        debug('clearInput called on unsupported backend');
        throw notImplementedError();
    }
}
class HelperProcessRDPBackendClient {
    async connect(config) {
        this.fatalHelperError = void 0;
        await this.ensureHelperStarted();
        const response = await this.send({
            type: 'connect',
            config
        });
        if ('connected' !== response.type) throw new Error(`Expected connected response, got ${response.type}`);
        this.connected = true;
        this.fatalHelperError = void 0;
        return response.info;
    }
    async disconnect() {
        const child = this.child;
        if (!child) return;
        let disconnectError;
        if (this.connected && null === child.exitCode) try {
            const response = await this.send({
                type: 'disconnect'
            });
            this.expectOk(response, 'disconnect');
        } catch (error) {
            disconnectError = error instanceof Error ? error : new Error(String(error));
        }
        this.connected = false;
        this.fatalHelperError = void 0;
        await this.shutdownHelper();
        if (disconnectError && !/RDP helper exited unexpectedly|RDP helper is not running|RDP helper shut down/u.test(disconnectError.message)) throw disconnectError;
    }
    async screenshotBase64() {
        const response = await this.send({
            type: 'screenshot'
        });
        if ('screenshot' !== response.type) throw new Error(`Expected screenshot response, got ${response.type}`);
        return response.base64;
    }
    async size() {
        const response = await this.send({
            type: 'size'
        });
        if ('size' !== response.type) throw new Error(`Expected size response, got ${response.type}`);
        return response.size;
    }
    async mouseMove(x, y) {
        const response = await this.send({
            type: 'mouseMove',
            x,
            y
        });
        this.expectOk(response, 'mouseMove');
    }
    async mouseButton(button, action) {
        const response = await this.send({
            type: 'mouseButton',
            button,
            action
        });
        this.expectOk(response, 'mouseButton');
    }
    async wheel(direction, amount, x, y) {
        const response = await this.send({
            type: 'wheel',
            direction,
            amount,
            x,
            y
        });
        this.expectOk(response, 'wheel');
    }
    async keyPress(keyName) {
        const response = await this.send({
            type: 'keyPress',
            keyName
        });
        this.expectOk(response, 'keyPress');
    }
    async typeText(text) {
        const response = await this.send({
            type: 'typeText',
            text
        });
        this.expectOk(response, 'typeText');
    }
    async clearInput() {
        const response = await this.send({
            type: 'clearInput'
        });
        this.expectOk(response, 'clearInput');
    }
    async ensureHelperStarted() {
        if (this.child && null === this.child.exitCode) return;
        const helperPath = this.resolveHelperPath();
        debug('starting rdp helper', {
            helperPath
        });
        const child = this.spawnFn(helperPath, [], {
            stdio: [
                'pipe',
                'pipe',
                'pipe'
            ]
        });
        child.stdout.setEncoding('utf8');
        child.stderr.setEncoding('utf8');
        this.child = child;
        this.stderrLines.length = 0;
        this.stdoutReader = createInterface({
            input: child.stdout,
            crlfDelay: 1 / 0
        });
        this.stderrReader = createInterface({
            input: child.stderr,
            crlfDelay: 1 / 0
        });
        this.stdoutReader.on('line', (line)=>{
            this.handleStdoutLine(line);
        });
        this.stderrReader.on('line', (line)=>{
            this.captureStderrLine(line);
        });
        child.on('exit', (code, signal)=>{
            this.connected = false;
            const error = this.createHelperError(`RDP helper exited unexpectedly (code=${code}, signal=${signal})`);
            this.fatalHelperError = error;
            this.rejectPending(error);
            this.disposeReaders();
            this.child = void 0;
        });
        child.on('error', (error)=>{
            this.connected = false;
            const helperError = this.createHelperError(`Failed to start RDP helper: ${error.message}`);
            this.fatalHelperError = helperError;
            this.rejectPending(helperError);
            this.disposeReaders();
            this.child = void 0;
        });
    }
    handleStdoutLine(line) {
        if (!line.trim()) return;
        let parsed;
        try {
            parsed = JSON.parse(line);
        } catch (error) {
            const protocolError = this.createHelperError(`RDP helper emitted malformed JSON: ${line}`);
            this.rejectPending(protocolError);
            this.shutdownHelper(protocolError);
            return;
        }
        const pending = this.pending.get(parsed.id);
        if (!pending) return void debug('dropping response for unknown request id', parsed);
        this.pending.delete(parsed.id);
        if (parsed.ok) return void pending.resolve(parsed.payload);
        pending.reject(this.createHelperError(parsed.error.message, parsed.error.code));
    }
    captureStderrLine(line) {
        if (!line.trim()) return;
        this.stderrLines.push(line);
        if (this.stderrLines.length > MAX_STDERR_LINES) this.stderrLines.shift();
    }
    async send(payload) {
        if ('connect' !== payload.type && this.fatalHelperError && (!this.child || null !== this.child.exitCode)) throw this.fatalHelperError;
        await this.ensureHelperStarted();
        const child = this.child;
        if (!child || null !== child.exitCode) throw this.createHelperError('RDP helper is not running');
        const id = `req-${++this.nextRequestId}`;
        const request = {
            id,
            payload
        };
        return new Promise((resolve, reject)=>{
            this.pending.set(id, {
                resolve,
                reject
            });
            child.stdin.write(`${JSON.stringify(request)}\n`, (error)=>{
                if (!error) return;
                this.pending.delete(id);
                reject(this.createHelperError(`Failed to send ${payload.type} request to RDP helper: ${error.message}`));
            });
        });
    }
    expectOk(response, actionName) {
        if ('ok' !== response.type) throw new Error(`Expected ok response for ${actionName}, got ${response.type}`);
    }
    rejectPending(error) {
        for (const { reject } of this.pending.values())reject(error);
        this.pending.clear();
    }
    createHelperError(message, code) {
        const stderrSummary = this.stderrLines.join('\n').trim();
        const suffix = stderrSummary ? `\nHelper stderr:\n${stderrSummary}` : '';
        const error = new Error(`${message}${suffix}`);
        if (code) error.name = code;
        return error;
    }
    disposeReaders() {
        this.stdoutReader?.close();
        this.stderrReader?.close();
        this.stdoutReader = void 0;
        this.stderrReader = void 0;
    }
    async shutdownHelper(rootError) {
        const child = this.child;
        this.child = void 0;
        this.disposeReaders();
        if (!child) return;
        this.rejectPending(rootError || this.createHelperError('RDP helper shut down'));
        if (null !== child.exitCode) return;
        child.stdin.end();
        const exited = Promise.race([
            once(child, 'exit'),
            new Promise((resolve)=>{
                setTimeout(()=>resolve('timeout'), HELPER_SHUTDOWN_TIMEOUT_MS);
            })
        ]);
        const result = await exited;
        if ('timeout' !== result) return;
        child.kill('SIGTERM');
        const terminated = Promise.race([
            once(child, 'exit'),
            new Promise((resolve)=>{
                setTimeout(()=>resolve('timeout'), HELPER_SHUTDOWN_TIMEOUT_MS);
            })
        ]);
        const terminateResult = await terminated;
        if ('timeout' !== terminateResult) return;
        child.kill('SIGKILL');
        await once(child, 'exit');
    }
    constructor(options){
        backend_client_define_property(this, "spawnFn", void 0);
        backend_client_define_property(this, "resolveHelperPath", void 0);
        backend_client_define_property(this, "child", void 0);
        backend_client_define_property(this, "stdoutReader", void 0);
        backend_client_define_property(this, "stderrReader", void 0);
        backend_client_define_property(this, "pending", new Map());
        backend_client_define_property(this, "stderrLines", []);
        backend_client_define_property(this, "nextRequestId", 0);
        backend_client_define_property(this, "connected", false);
        backend_client_define_property(this, "fatalHelperError", void 0);
        this.spawnFn = options?.spawnFn || spawn;
        const overridePath = options?.helperPath;
        this.resolveHelperPath = overridePath ? ()=>overridePath : getRdpHelperBinaryPath;
    }
}
function createDefaultRDPBackendClient() {
    return new HelperProcessRDPBackendClient();
}
function device_define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
const device_debug = getDebug('rdp:device');
const device_SMOOTH_MOVE_STEPS_TAP = 8;
const device_SMOOTH_MOVE_STEPS_MOUSE_MOVE = 10;
const SMOOTH_MOVE_STEPS_DRAG = 12;
const device_SMOOTH_MOVE_DELAY_TAP = 8;
const device_SMOOTH_MOVE_DELAY_MOUSE_MOVE = 10;
const SMOOTH_MOVE_DELAY_DRAG = 10;
const device_MOUSE_MOVE_EFFECT_WAIT = 300;
const device_CLICK_HOLD_DURATION = 50;
const DRAG_HOLD_DURATION = 100;
const device_INPUT_FOCUS_DELAY = 300;
const device_INPUT_CLEAR_DELAY = 150;
const device_SCROLL_STEP_DELAY = 100;
const device_SCROLL_COMPLETE_DELAY = 500;
const DEFAULT_SCROLL_DISTANCE = 480;
const device_EDGE_SCROLL_STEPS = 10;
const DEFAULT_SCROLL_STEP_AMOUNT = 120;
class RDPDevice {
    describe() {
        const port = this.options.port || 3389;
        const username = this.options.username ? ` as ${this.options.username}` : '';
        const session = this.connectionInfo?.sessionId ? ` [session ${this.connectionInfo.sessionId}]` : '';
        return `RDP Device ${this.options.host}:${port}${username}${session}`;
    }
    async connect() {
        this.throwIfDestroyed();
        device_debug('connecting to rdp backend', {
            host: this.options.host,
            port: this.options.port,
            username: this.options.username
        });
        this.connectionInfo = await this.backend.connect(this.options);
        this.cursorPosition = [
            Math.round(this.connectionInfo.size.width / 2),
            Math.round(this.connectionInfo.size.height / 2)
        ];
    }
    async screenshotBase64() {
        this.assertConnected();
        return this.backend.screenshotBase64();
    }
    async size() {
        this.assertConnected();
        return this.backend.size();
    }
    async destroy() {
        if (this.destroyed) return;
        this.destroyed = true;
        this.connectionInfo = void 0;
        this.cursorPosition = void 0;
        await this.backend.disconnect();
    }
    actionSpace() {
        const defaultActions = [
            defineActionTap(async ({ locate })=>{
                const element = this.requireLocate(locate, 'tap');
                await this.moveToElement(element, {
                    steps: device_SMOOTH_MOVE_STEPS_TAP,
                    stepDelayMs: device_SMOOTH_MOVE_DELAY_TAP
                });
                await this.backend.mouseButton('left', 'down');
                await sleep(device_CLICK_HOLD_DURATION);
                await this.backend.mouseButton('left', 'up');
            }),
            defineActionDoubleClick(async ({ locate })=>{
                const element = this.requireLocate(locate, 'double click');
                await this.moveToElement(element, {
                    steps: device_SMOOTH_MOVE_STEPS_TAP,
                    stepDelayMs: device_SMOOTH_MOVE_DELAY_TAP
                });
                await this.backend.mouseButton('left', 'doubleClick');
            }),
            defineActionRightClick(async ({ locate })=>{
                const element = this.requireLocate(locate, 'right click');
                await this.moveToElement(element, {
                    steps: device_SMOOTH_MOVE_STEPS_TAP,
                    stepDelayMs: device_SMOOTH_MOVE_DELAY_TAP
                });
                await this.backend.mouseButton('right', 'click');
            }),
            defineActionHover(async ({ locate })=>{
                const element = this.requireLocate(locate, 'hover');
                await this.moveToElement(element, {
                    steps: device_SMOOTH_MOVE_STEPS_MOUSE_MOVE,
                    stepDelayMs: device_SMOOTH_MOVE_DELAY_MOUSE_MOVE,
                    settleDelayMs: device_MOUSE_MOVE_EFFECT_WAIT
                });
            }),
            defineActionInput(async (param)=>{
                this.assertConnected();
                if (param.locate) {
                    await this.moveToElement(param.locate, {
                        steps: device_SMOOTH_MOVE_STEPS_TAP,
                        stepDelayMs: device_SMOOTH_MOVE_DELAY_TAP
                    });
                    await this.backend.mouseButton('left', 'click');
                    await sleep(device_INPUT_FOCUS_DELAY);
                }
                if ('typeOnly' !== param.mode) {
                    await this.clearInput();
                    await sleep(device_INPUT_CLEAR_DELAY);
                }
                if ('clear' === param.mode) return;
                if (param.value) await this.backend.typeText(param.value);
            }),
            defineActionClearInput(async ({ locate })=>{
                this.assertConnected();
                if (locate) {
                    await this.moveToElement(locate, {
                        steps: device_SMOOTH_MOVE_STEPS_TAP,
                        stepDelayMs: device_SMOOTH_MOVE_DELAY_TAP
                    });
                    await this.backend.mouseButton('left', 'click');
                    await sleep(device_INPUT_FOCUS_DELAY);
                }
                await this.clearInput();
                await sleep(device_INPUT_CLEAR_DELAY);
            }),
            defineActionKeyboardPress(async ({ locate, keyName })=>{
                this.assertConnected();
                if (locate) {
                    await this.moveToElement(locate, {
                        steps: device_SMOOTH_MOVE_STEPS_TAP,
                        stepDelayMs: device_SMOOTH_MOVE_DELAY_TAP
                    });
                    await this.backend.mouseButton('left', 'click');
                }
                await this.backend.keyPress(keyName);
            }),
            defineActionScroll(async (param)=>{
                this.assertConnected();
                const target = param.locate;
                if (target) await this.moveToElement(target, {
                    steps: device_SMOOTH_MOVE_STEPS_MOUSE_MOVE,
                    stepDelayMs: device_SMOOTH_MOVE_DELAY_MOUSE_MOVE
                });
                if (param.scrollType && 'singleAction' !== param.scrollType) {
                    const direction = this.edgeScrollDirection(param.scrollType);
                    for(let i = 0; i < device_EDGE_SCROLL_STEPS; i++)await this.performWheel(direction, DEFAULT_SCROLL_DISTANCE, target?.center[0], target?.center[1]);
                    await sleep(device_SCROLL_COMPLETE_DELAY);
                    return;
                }
                await this.performWheel(param.direction || 'down', param.distance || DEFAULT_SCROLL_DISTANCE, target?.center[0], target?.center[1]);
                await sleep(device_SCROLL_COMPLETE_DELAY);
            }),
            defineActionDragAndDrop(async ({ from, to })=>{
                this.assertConnected();
                const source = this.requireLocate(from, 'drag source');
                const target = this.requireLocate(to, 'drag target');
                await this.moveToElement(source, {
                    steps: device_SMOOTH_MOVE_STEPS_TAP,
                    stepDelayMs: device_SMOOTH_MOVE_DELAY_TAP
                });
                await this.backend.mouseButton('left', 'down');
                await sleep(DRAG_HOLD_DURATION);
                await this.moveToElement(target, {
                    steps: SMOOTH_MOVE_STEPS_DRAG,
                    stepDelayMs: SMOOTH_MOVE_DELAY_DRAG
                });
                await sleep(DRAG_HOLD_DURATION);
                await this.backend.mouseButton('left', 'up');
            }),
            defineAction({
                name: 'MiddleClick',
                description: 'Middle click the element',
                sample: {
                    locate: {
                        prompt: 'the browser tab close target'
                    }
                },
                paramSchema: actionTapParamSchema,
                call: async ({ locate })=>{
                    const element = this.requireLocate(locate, 'middle click');
                    await this.moveToElement(element, {
                        steps: device_SMOOTH_MOVE_STEPS_TAP,
                        stepDelayMs: device_SMOOTH_MOVE_DELAY_TAP
                    });
                    await this.backend.mouseButton('middle', 'click');
                }
            }),
            defineAction({
                name: 'ListDisplays',
                description: 'List all available displays/monitors',
                call: async ()=>{
                    this.assertConnected();
                    const size = await this.size();
                    return [
                        {
                            id: this.connectionInfo?.sessionId || this.options.host,
                            name: `RDP ${this.connectionInfo?.server || this.options.host} (${size.width}x${size.height})`,
                            primary: true
                        }
                    ];
                }
            })
        ];
        return [
            ...defaultActions,
            ...this.options.customActions || []
        ];
    }
    assertConnected() {
        this.throwIfDestroyed();
        if (!this.connectionInfo) throw new Error('RDPDevice is not connected');
    }
    throwIfDestroyed() {
        if (this.destroyed) throw new Error('RDPDevice has been destroyed');
    }
    requireLocate(locate, actionName) {
        if (!locate) throw new Error(`Missing target element for ${actionName}`);
        return locate;
    }
    async moveToElement(element, options) {
        this.assertConnected();
        const targetX = Math.round(element.center[0]);
        const targetY = Math.round(element.center[1]);
        await this.movePointer(targetX, targetY, options);
    }
    async clearInput() {
        if (this.backend.clearInput) return void await this.backend.clearInput();
        await this.backend.keyPress('Control+A');
        await this.backend.keyPress('Backspace');
    }
    edgeScrollDirection(scrollType) {
        switch(scrollType){
            case 'scrollToTop':
                return 'up';
            case 'scrollToBottom':
                return 'down';
            case 'scrollToLeft':
                return 'left';
            case 'scrollToRight':
                return 'right';
            case 'singleAction':
                return 'down';
            default:
                throw new Error(`Unsupported scroll type: ${scrollType}`);
        }
    }
    async movePointer(targetX, targetY, options) {
        this.assertConnected();
        const start = this.cursorPosition || [
            targetX,
            targetY
        ];
        const steps = Math.max(1, options?.steps || 1);
        const stepDelayMs = options?.stepDelayMs || 0;
        for(let step = 1; step <= steps; step++){
            const x = Math.round(start[0] + (targetX - start[0]) * step / steps);
            const y = Math.round(start[1] + (targetY - start[1]) * step / steps);
            await this.backend.mouseMove(x, y);
            this.cursorPosition = [
                x,
                y
            ];
            if (stepDelayMs > 0 && step < steps) await sleep(stepDelayMs);
        }
        if (options?.settleDelayMs) await sleep(options.settleDelayMs);
    }
    async performWheel(direction, amount, x, y) {
        let remaining = Math.abs(amount);
        if (0 === remaining) remaining = DEFAULT_SCROLL_STEP_AMOUNT;
        while(remaining > 0){
            const chunk = Math.min(remaining, DEFAULT_SCROLL_STEP_AMOUNT);
            await this.backend.wheel(direction, chunk, x, y);
            remaining -= chunk;
            if (remaining > 0) await sleep(device_SCROLL_STEP_DELAY);
        }
    }
    constructor(options){
        device_define_property(this, "interfaceType", 'rdp');
        device_define_property(this, "options", void 0);
        device_define_property(this, "backend", void 0);
        device_define_property(this, "connectionInfo", void 0);
        device_define_property(this, "destroyed", false);
        device_define_property(this, "cursorPosition", void 0);
        device_define_property(this, "uri", void 0);
        this.options = {
            port: 3389,
            securityProtocol: 'auto',
            ignoreCertificate: false,
            ...options
        };
        this.backend = options.backend || createDefaultRDPBackendClient();
    }
}
class ComputerAgent extends Agent {
}
function createLocalComputerDevice(opts) {
    return new ComputerDevice({
        displayId: opts?.displayId,
        customActions: opts?.customActions,
        keyboardDriver: opts?.keyboardDriver,
        headless: opts?.headless,
        xvfbResolution: opts?.xvfbResolution
    });
}
function createRDPComputerDevice(opts) {
    return new RDPDevice({
        host: opts.host,
        port: opts.port,
        username: opts.username,
        password: opts.password,
        domain: opts.domain,
        adminSession: opts.adminSession,
        ignoreCertificate: opts.ignoreCertificate,
        securityProtocol: opts.securityProtocol,
        desktopWidth: opts.desktopWidth,
        desktopHeight: opts.desktopHeight,
        backend: opts.backend,
        customActions: opts.customActions
    });
}
async function agentFromComputer(opts) {
    const device = createLocalComputerDevice(opts);
    await device.connect();
    return new ComputerAgent(device, opts);
}
async function agentForRDPComputer(opts) {
    const device = createRDPComputerDevice(opts);
    await device.connect();
    return new ComputerAgent(device, opts);
}
function mcp_tools_define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
const mcp_tools_debug = getDebug('mcp:computer-tools');
const computerInitArgShape = {
    displayId: z.string().optional().describe('Display ID (from computer_list_displays)'),
    headless: z.boolean().optional().describe('Start virtual display via Xvfb (Linux only)')
};
class ComputerMidsceneTools extends BaseMidsceneTools {
    getCliReportSessionName() {
        return 'midscene-computer';
    }
    createTemporaryDevice() {
        return new ComputerDevice({});
    }
    async ensureAgent(opts) {
        const displayId = opts?.displayId;
        const headless = opts?.headless;
        if (this.agent && (void 0 !== displayId || void 0 !== headless)) {
            try {
                await this.agent.destroy?.();
            } catch (error) {
                mcp_tools_debug('Failed to destroy agent during cleanup:', error);
            }
            this.agent = void 0;
        }
        if (this.agent) return this.agent;
        mcp_tools_debug('Creating Computer agent with displayId:', displayId || 'primary');
        const reportOptions = this.readCliReportAgentOptions();
        const agentOpts = {
            ...displayId ? {
                displayId
            } : {},
            ...void 0 !== headless ? {
                headless
            } : {},
            ...reportOptions ?? {}
        };
        const agent = await agentFromComputer(Object.keys(agentOpts).length > 0 ? agentOpts : void 0);
        this.agent = agent;
        return agent;
    }
    preparePlatformTools() {
        return [
            {
                name: 'computer_connect',
                description: 'Connect to computer desktop. Provide displayId to connect to a specific display (use computer_list_displays to get available IDs). If not provided, uses the primary display.',
                schema: this.getAgentInitArgSchema(),
                cli: this.getAgentInitArgCliMetadata(),
                handler: async (args)=>{
                    const initArgs = this.extractAgentInitParam(args);
                    const reportSession = this.createNewCliReportSession(initArgs?.displayId ?? 'primary');
                    this.commitCliReportSession(reportSession);
                    if (this.agent) {
                        try {
                            await this.agent.destroy?.();
                        } catch (error) {
                            mcp_tools_debug('Failed to destroy agent during connect:', error);
                        }
                        this.agent = void 0;
                    }
                    const agent = await this.ensureAgent(initArgs);
                    const screenshot = await agent.interface.screenshotBase64();
                    return {
                        content: [
                            {
                                type: 'text',
                                text: `Connected to computer${initArgs?.displayId ? ` (Display: ${initArgs.displayId})` : ' (Primary display)'}`
                            },
                            ...this.buildScreenshotContent(screenshot)
                        ]
                    };
                }
            },
            {
                name: 'computer_disconnect',
                description: 'Disconnect from computer and release resources',
                schema: {},
                handler: this.createDisconnectHandler('computer')
            },
            {
                name: 'computer_list_displays',
                description: 'List all available displays/monitors',
                schema: {},
                handler: async ()=>{
                    const displays = await ComputerDevice.listDisplays();
                    return {
                        content: [
                            {
                                type: 'text',
                                text: `Available displays:\n${displays.map((d)=>`- ${d.name} (ID: ${d.id})${d.primary ? ' [PRIMARY]' : ''}`).join('\n')}`
                            }
                        ]
                    };
                }
            }
        ];
    }
    constructor(...args){
        super(...args), mcp_tools_define_property(this, "initArgSpec", {
            namespace: 'computer',
            shape: computerInitArgShape,
            cli: {
                preferBareKeys: true
            },
            adapt: (extracted)=>extracted
        });
    }
}
function version() {
    const currentVersion = "1.7.9";
    console.log(`@midscene/computer v${currentVersion}`);
    return currentVersion;
}
function checkAccessibilityPermission(promptIfNeeded = false) {
    if ('darwin' !== process.platform) return {
        hasPermission: true,
        platform: process.platform
    };
    try {
        let permissions;
        try {
            const dynamicRequire = createRequire(import.meta.url);
            permissions = dynamicRequire('node-mac-permissions');
        } catch  {
            return {
                hasPermission: true,
                platform: process.platform
            };
        }
        const status = permissions.getAuthStatus('accessibility');
        if ('authorized' === status) return {
            hasPermission: true,
            platform: process.platform
        };
        if (promptIfNeeded) permissions.askForAccessibilityAccess();
        return {
            hasPermission: false,
            platform: process.platform,
            error: `macOS Accessibility permission is required (current status: ${status}).\n\nPlease follow these steps:\n1. Open System Settings > Privacy & Security > Accessibility\n2. Enable the application running this script (e.g., Terminal, iTerm2, VS Code, WebStorm)\n3. Restart your terminal or IDE after granting permission\n\nFor more details, see: https://github.com/nut-tree/nut.js#macos`
        };
    } catch (error) {
        return {
            hasPermission: false,
            platform: process.platform,
            error: `Failed to check accessibility permission: ${error instanceof Error ? error.message : String(error)}`
        };
    }
}
async function checkComputerEnvironment() {
    try {
        const libnutModule = await import("@computer-use/libnut/dist/import_libnut.js");
        const libnut = libnutModule.libnut;
        const screenSize = libnut.getScreenSize();
        if (!screenSize || screenSize.width <= 0) return {
            available: false,
            error: 'libnut cannot get screen size',
            platform: process.platform,
            displays: 0
        };
        const displays = await ComputerDevice.listDisplays();
        return {
            available: true,
            platform: process.platform,
            displays: displays.length
        };
    } catch (error) {
        const errorMessage = error instanceof Error ? error.message : String(error);
        return {
            available: false,
            error: errorMessage,
            platform: process.platform,
            displays: 0
        };
    }
}
async function getConnectedDisplays() {
    return ComputerDevice.listDisplays();
}
export { ComputerAgent, ComputerDevice, ComputerMidsceneTools, HelperProcessRDPBackendClient, RDPDevice, UnsupportedRDPBackendClient, agentForRDPComputer, agentFromComputer, checkAccessibilityPermission, checkComputerEnvironment, checkXvfbInstalled, createDefaultRDPBackendClient, getConnectedDisplays, needsXvfb, overrideAIConfig, version };
